# Genomics Pipelines  
### Description:  
Automated pipelines, scripts, and command-line operations to perform data processing, analysis, and visualizations of genomics data. 

- Process NGS .fastq.gz raw reads using an automated script to produce .bam, .bw, and .bed files. Executed on a high-performance computing platform.  
- Perform downstream analysis such as differential gene expression, gene annotations, and motif discovery.  
- Analyze single-cell RNA-Seq data by clustering, differential gene expression, and trajectory inference.  
  
### Prerequisites:    
- Linux terminal/bash/shell and command-line usage  
- Python and conda/virtualenv package management  
- R and RStudio  
- Jupyter Notebook  
- Slurm job scheduling  
  
### Resources  
CUT&Tag https://www.nature.com/articles/s41467-019-09982-5  
Pipeline example https://yezhengstat.github.io/CUTTag_tutorial/  
Additional https://learn.gencore.bio.nyu.edu/  
   

___  
 

