# Genomics Pipelines  
### Description:  
Automated pipelines, scripts, and command-line operations to perform data processing, analysis, and visualizations of genomics data. 

- Process NGS .fastq.gz raw reads using an automated script to produce .bam, .bw, and .bed files. Executed on a high-performance computing platform.  
- Perform downstream analysis such as differential gene expression, gene annotations, and motif discovery.  
- Analyze single-cell RNA-Seq data by clustering, differential gene expression, and trajectory inference.  
  
### Prerequisites:    
- Linux terminal/bash/shell and <a href="https://ubuntu.com/tutorials/command-line-for-beginners#1-overview">command-line interface usage</a>  
- <a href="https://www.python.org/">Python</a> and <a href="https://docs.conda.io/en/latest/">conda</a>/<a href="https://virtualenv.pypa.io/en/latest/">virtualenv</a> package environment management  
- <a href="https://www.r-project.org/">R</a> and <a href="https://posit.co/downloads/">RStudio</a>  
- <a href="https://docs.jupyter.org/en/latest/">Jupyter Notebook</a>  
- <a href="https://slurm.schedmd.com/overview.html">Slurm</a> job scheduling  
  
### Resources  
NGS Analysis https://learn.gencore.bio.nyu.edu/  
   

___  
 

