{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f10c1ed7",
   "metadata": {},
   "source": [
    "# <center> RNA-Seq Differential Gene Expression Notebook </center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa91750",
   "metadata": {},
   "source": [
    "## <center>Preface</center>  \n",
    "<p><center>This Notebook provides an interactive way to process, analyze, and view differential gene expressions from next-generation sequencing data.<br>The following steps serve as a guide, however, care should be taken when executing and interpreting results.</center><br>This notebook uses the following R packages: <li> - <a href=\"https://bioconductor.org/packages/release/bioc/html/Rsubread.html\"><strong>Rsubread</strong></a> to produce a count matrix from provided .bam files using featureCounts (if matrix not already created). <br>- <a href=\"https://bioconductor.org/packages/release/bioc/html/DESeq2.html\"><strong>DESeq2</strong></a> to process the count matrix and produce differential gene expression statistics. <br>- <a href=\"https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html\"><strong>clusterProfiler</strong></a> to annotate genes (GO, GSEA, KEGG) using database packages for mappings. </li><center><br><em>Note that despite following the framework below, it is up to <strong>you</strong> to determine which steps or values to change or ignore based on your knowledge of experimental details in obtaining data and your analysis goals.</em></center></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0563cf6",
   "metadata": {},
   "source": [
    "### sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ef8bc78",
   "metadata": {},
   "source": [
    "R version 4.2.2 Patched (2022-11-10 r83330)\n",
    "Platform: x86_64-pc-linux-gnu (64-bit)\n",
    "Running under: Ubuntu 22.04.1 LTS\n",
    "\n",
    "Matrix products: default\n",
    "BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0\n",
    "LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0\n",
    "\n",
    "locale:\n",
    " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
    " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
    " [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
    "\n",
    "attached base packages:\n",
    "[1] stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     \n",
    "\n",
    "other attached packages:\n",
    " [1] optparse_1.7.3                            org.Mm.eg.db_3.16.0                      \n",
    " [3] org.Hs.eg.db_3.16.0                       TxDb.Hsapiens.UCSC.hg38.knownGene_3.16.0 \n",
    " [5] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2   TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0\n",
    " [7] GenomicFeatures_1.50.3                    AnnotationDbi_1.60.0                     \n",
    " [9] ggplot2_3.4.0                             clusterProfiler_4.6.0                    \n",
    "[11] DESeq2_1.38.3                             SummarizedExperiment_1.28.0              \n",
    "[13] Biobase_2.58.0                            MatrixGenerics_1.10.0                    \n",
    "[15] matrixStats_0.63.0                        GenomicRanges_1.50.2                     \n",
    "[17] GenomeInfoDb_1.34.6                       IRanges_2.32.0                           \n",
    "[19] S4Vectors_0.36.1                          Rsubread_2.12.2                          \n",
    "[21] Rgraphviz_2.42.0                          graph_1.76.0                             \n",
    "[23] BiocGenerics_0.44.0                      \n",
    "\n",
    "loaded via a namespace (and not attached):\n",
    "  [1] shadowtext_0.1.2         fastmatch_1.1-3          BiocFileCache_2.6.0      plyr_1.8.8              \n",
    "  [5] igraph_1.3.5             lazyeval_0.2.2           splines_4.2.2            BiocParallel_1.32.5     \n",
    "  [9] digest_0.6.31            yulab.utils_0.0.6        GOSemSim_2.24.0          viridis_0.6.2           \n",
    " [13] GO.db_3.16.0             fansi_1.0.4              magrittr_2.0.3           memoise_2.0.1           \n",
    " [17] Biostrings_2.66.0        annotate_1.76.0          graphlayouts_0.8.4       enrichplot_1.18.3       \n",
    " [21] prettyunits_1.1.1        colorspace_2.1-0         blob_1.2.3               rappdirs_0.3.3          \n",
    " [25] ggrepel_0.9.2            dplyr_1.0.10             crayon_1.5.2             RCurl_1.98-1.9          \n",
    " [29] jsonlite_1.8.4           scatterpie_0.1.8         ape_5.6-2                glue_1.6.2              \n",
    " [33] polyclip_1.10-4          gtable_0.3.1             zlibbioc_1.44.0          XVector_0.38.0          \n",
    " [37] DelayedArray_0.24.0      scales_1.2.1             DOSE_3.24.2              DBI_1.1.3               \n",
    " [41] Rcpp_1.0.10              viridisLite_0.4.1        xtable_1.8-4             progress_1.2.2          \n",
    " [45] gridGraphics_0.5-1       tidytree_0.4.2           bit_4.0.5                getopt_1.20.3           \n",
    " [49] httr_1.4.4               fgsea_1.24.0             RColorBrewer_1.1-3       ellipsis_0.3.2          \n",
    " [53] pkgconfig_2.0.3          XML_3.99-0.13            farver_2.1.1             dbplyr_2.3.0            \n",
    " [57] locfit_1.5-9.7           utf8_1.2.2               ggplotify_0.1.0          tidyselect_1.2.0        \n",
    " [61] rlang_1.0.6              reshape2_1.4.4           munsell_0.5.0            tools_4.2.2             \n",
    " [65] cachem_1.0.6             downloader_0.4           cli_3.6.0                generics_0.1.3          \n",
    " [69] RSQLite_2.2.20           gson_0.0.9               stringr_1.5.0            fastmap_1.1.0           \n",
    " [73] yaml_2.3.7               ggtree_3.6.2             bit64_4.0.5              tidygraph_1.2.2         \n",
    " [77] purrr_1.0.1              KEGGREST_1.38.0          ggraph_2.1.0             nlme_3.1-161            \n",
    " [81] aplot_0.1.9              xml2_1.3.3               biomaRt_2.54.0           compiler_4.2.2          \n",
    " [85] rstudioapi_0.14          filelock_1.0.2           curl_5.0.0               png_0.1-8               \n",
    " [89] treeio_1.22.0            tibble_3.1.8             tweenr_2.0.2             geneplotter_1.76.0      \n",
    " [93] stringi_1.7.12           lattice_0.20-45          Matrix_1.5-1             vctrs_0.5.2             \n",
    " [97] pillar_1.8.1             lifecycle_1.0.3          data.table_1.14.6        cowplot_1.1.1           \n",
    "[101] bitops_1.0-7             patchwork_1.1.2          rtracklayer_1.58.0       qvalue_2.30.0           \n",
    "[105] R6_2.5.1                 BiocIO_1.8.0             gridExtra_2.3            codetools_0.2-18        \n",
    "[109] MASS_7.3-58              assertthat_0.2.1         rjson_0.2.21             withr_2.5.0             \n",
    "[113] GenomicAlignments_1.34.0 Rsamtools_2.14.0         GenomeInfoDbData_1.2.9   parallel_4.2.2          \n",
    "[117] hms_1.1.2                ggfun_0.0.9              tidyr_1.3.0              HDO.db_0.99.1           \n",
    "[121] ggforce_0.4.1            restfulr_0.0.15 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f24f257",
   "metadata": {},
   "source": [
    "### Required:\n",
    "<br><strong>1) A <u>COUNT MATRIX FILE</u> produced by featureCounts and/or htseq-count which should be comma-separated (.csv), example formatted as follows:</strong><br>\n",
    "<table style=’float:left;’>\n",
    "    <tr>\n",
    "        <th>GeneID</th>\n",
    "        <th>BamSample 1</th>\n",
    "        <th>BamSample 2</th>\n",
    "        <th>BamSample 3</th>\n",
    "        <th>BamSample N</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene1</td>\n",
    "        <td>123</td>\n",
    "        <td>456</td>\n",
    "        <td>78</td>\n",
    "        <td>89</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene2</td>\n",
    "        <td>321</td>\n",
    "        <td>654</td>\n",
    "        <td>45</td>\n",
    "        <td>12</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>gene3</td>\n",
    "        <td>456</td>\n",
    "        <td>741</td>\n",
    "        <td>0</td>\n",
    "        <td>0</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>geneN</td>\n",
    "        <td>369</td>\n",
    "        <td>987</td>\n",
    "        <td>100</td>\n",
    "        <td>200</td>\n",
    "    </tr>\n",
    "</table><br>\n",
    "&emsp;<strong>There should be no duplicated gene IDs in the first column.</strong><br>\n",
    "&emsp;<strong>---This file may be produced using this notebook with the provided .bam files.---</strong>\n",
    "<br><br>\n",
    "<strong>2) A <u>SAMPLE INFO FILE</u> is also required and should be comma-separated (.csv), example formatted as follows:</strong><br>\n",
    "<table style=’float:left;’>\n",
    "    <tr>\n",
    "        <th>Sample</th>\n",
    "        <th>Condition</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>BamSample 1</td>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>BamSample 2</td>\n",
    "        <td>control</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>BamSample 3</td>\n",
    "        <td>experimental</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>BamSample N</td>\n",
    "        <td>experimental</td>\n",
    "    </tr>\n",
    "</table><br>\n",
    "&emsp;<strong>Must contain a column with the header named \"Condition\" for grouping samples.</strong><br>\n",
    "&emsp;<strong>There must be more than 1 sample for each condition.</strong><br>\n",
    "&emsp;<strong>Row order does not matter.</strong><br>\n",
    "&emsp;<strong>May contain multiple conditions. Pair-wise comparisons are performed for differential gene expressions between each pair.</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2964963f",
   "metadata": {},
   "source": [
    "## <center> ======= Load Packages ======= </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e76db7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "  library(Rsubread)\n",
    "  library(DESeq2)\n",
    "  library(clusterProfiler)\n",
    "  library(ggplot2)\n",
    "  library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
    "  library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "  library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "  #library(ensembldb)\n",
    "  #library(EnsDb.Hsapiens.v86)\n",
    "  #library(EnsDb.Mmusculus.v79)\n",
    "  library(org.Hs.eg.db)\n",
    "  library(org.Mm.eg.db)\n",
    "  library(optparse)\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8a525e",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85d9e80e",
   "metadata": {},
   "source": [
    "### <center> Generate Count Matrix <br>(Skip if already provided)</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37fff17e",
   "metadata": {},
   "source": [
    "###  First, put all .bam (and .bai) files into a directory, e.g. \"bams/\", then run the following cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3571555",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in bam filenames from directory path\n",
    "bam_files <- list.files(path=\"bams/\", full.names=TRUE)[c(TRUE, FALSE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb9a67b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count gene alignments (select appropriate assembly, mm10, hg38, etc...and check featureCount parameters) \n",
    "bamcounts <- featureCounts(bam_files, annot.inbuilt=\"mm10\", ignoreDup=TRUE, isPairedEnd=TRUE, nthreads=4, verbose=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c4face4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map gene NCBI Entrez IDs to Gene Symbols\n",
    "rownames(bamcounts$counts) <- mapIds(org.Mm.eg.db, keys = rownames(bamcounts$counts), column = \"SYMBOL\", keytype = \"ENTREZID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1763eb86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop any unmapped genes\n",
    "bamcounts$counts <- bamcounts$counts[!(is.na(rownames(bamcounts$counts))), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515083e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write featureCounts output to file in current directory\n",
    "for (n in names(bamcounts)){\n",
    "  write.table(bamcounts[[n]], file=paste(getwd(), \"/\", n, \".csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7081102",
   "metadata": {},
   "source": [
    "### Check output files to ensure counts matrix was successfully created before proceeding."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a3dfee",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf056baf",
   "metadata": {},
   "source": [
    "## <center> ======= Define Variables ======= </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f31e366",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ======= Get command-line optional arguments (defaults) =======\n",
    "option_list = list(\n",
    "  make_option(c(\"-c\", \"--countsfile\"), type=\"character\", help=\"Count matrix file\", metavar=\"character\"),\n",
    "  make_option(c(\"-s\", \"--sampleinfo\"), type=\"character\", help=\"File containing samples and their experimental conditions info (row names must match matrix file column names)\", metavar=\"character\"),\n",
    "  make_option(c(\"-o\", \"--organism\"), type=\"character\", default=\"mouse\", help=\"Organism to annotate genes (mouse or human)\", metavar=\"character\"),\n",
    "  make_option(c(\"-r\", \"--result_dir\"), type=\"character\", default=\"DEG_Analysis/\", help=\"Directory name for saving output results\", metavar=\"character\"),\n",
    "  make_option(c(\"-f\", \"--filter\"), action=\"store_true\", type=\"logical\", default=TRUE, help=\"Flag to filter read counts (removes 0 counts, then removes genes in bottom 10th percentile of counts for those in < 3 samples)\", metavar=\"character\"),\n",
    "  make_option(c(\"--min_count\"), type=\"integer\", default=1, help=\"Minimum counts to include if filtering\", metavar=\"integer\"),\n",
    "  make_option(c(\"--lfc\"), type=\"double\", default=1.5, help=\"Magnitude of log2foldchange to define significant up/down regulation of genes\", metavar=\"integer\")\n",
    "  );\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d81a18a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to count matrix file\n",
    "opt$countsfile <- \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9922e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to sample info file\n",
    "opt$sampleinfo <- \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a221e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View variables\n",
    "for (i in which(names(opt) != \"help\")) {\n",
    "  cat(names(opt)[i], '=', paste(opt)[i], \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1321a63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually re-define any values if defaults not desired, otherwise SKIP THIS CELL\n",
    "opt$countsfile <- \"\"\n",
    "opt$sampleinfo <- \"\"\n",
    "opt$organism <- \"\"\n",
    "opt$result_dir <- \"DEG_Analysis/\"\n",
    "opt$filter <- TRUE # or FALSE\n",
    "opt$min_count <- 1 # integer value\n",
    "opt$lfc <- 1.5 # any numerical value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2403cc6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define references for annotations\n",
    "if (tolower(opt$organism) == \"mouse\"){\n",
    "  annoDb <- \"org.Mm.eg.db\"\n",
    "  keggOrg <- \"mmu\"\n",
    "} else if (tolower(opt$organism) == \"human\"){\n",
    "  annoDb <- \"org.Hs.eg.db\"\n",
    "  keggOrg <- \"hsa\"\n",
    "} else{\n",
    "  stop(\"Invalid choice of organism\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b9a8166",
   "metadata": {},
   "source": [
    "## <center> =========== Load Input Files and Prep Run ============ </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2621c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_mtx <- as.matrix(read.csv(opt$countsfile, sep=\",\", row.names=1, check.names=FALSE))\n",
    "# View \n",
    "head(count_mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5380ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleinfo <- read.csv(opt$sampleinfo, row.names=1)\n",
    "# View\n",
    "sampleinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc50ee69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create output file directory and set as working directory\n",
    "if (!file.exists(opt$result_dir)) {\n",
    "  dir.create(opt$result_dir)\n",
    "}\n",
    "setwd(opt$result_dir)\n",
    "cat(\"Output files will be in\", paste(getwd(), \"/\", sep=\"\"), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9426556b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all pairwise combinations of conditions for comparing (column values shown below)\n",
    "combs <- as.data.frame(combn(unique(sampleinfo$Condition), 2))\n",
    "combs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "573b83c6",
   "metadata": {},
   "source": [
    "## <center> =========== Run Analysis ============ </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c147385",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through condition combinations to compare DEG\n",
    "for (c in colnames(combs)){\n",
    "  \n",
    "  comparison <- paste(combs[[c]][1], combs[[c]][2], sep=\"_vs_\")\n",
    "  cat(\"\\n==========\", comparison, \"==========\\n\")\n",
    "  # Isolate matrix for relevant samples\n",
    "  samples1 <- rownames(sampleinfo[sampleinfo$Condition == combs[[c]][1], ])\n",
    "  samples2 <- rownames(sampleinfo[sampleinfo$Condition == combs[[c]][2], ])\n",
    "  mtx <- count_mtx[, c(samples1, samples2)]\n",
    "  mtx_info <- sampleinfo[c(samples1, samples2), ]\n",
    "  mtx_info$Condition <- factor(mtx_info$Condition)\n",
    "  \n",
    "  # Check that matrix columns are same order as sample info\n",
    "  if (!all(rownames(mtx_info) == colnames(mtx))) {\n",
    "    break\n",
    "  }\n",
    "  \n",
    "  # Create DESeqDataSet object from input\n",
    "  dds <- DESeqDataSetFromMatrix(countData = mtx,\n",
    "                                colData = mtx_info,\n",
    "                                design = ~ Condition)\n",
    "  \n",
    "  # Create directory for output files of current comparison\n",
    "  if (!file.exists(comparison)) {\n",
    "    dir.create(comparison)\n",
    "  }\n",
    "  \n",
    "  # Set output filenames prefix\n",
    "  output_prefix <- paste(getwd(), \"/\", comparison, \"/\", sep=\"\")\n",
    "  \n",
    "  cat(\"Genes and samples:\\n\", dim(dds), \"\\n\")\n",
    "  \n",
    "  # =========== Filter Count Matrix ============\n",
    "  if (opt$filter) {\n",
    "    \n",
    "    # Get percentile ranges of counts\n",
    "    cat(\"Percentiles of genes count sums over all samples:\\n\")\n",
    "    cat(quantile(rowSums(counts(dds)), probs = c(0, 0.1, 0.25, 0.5, 0.75, 0.9, 1)), \"\\n\")\n",
    "    \n",
    "    # Remove 0 counts to avoid 0-bias and disregard irrelevant genes\n",
    "    keep <- rowSums(counts(dds)) >= opt$min_count\n",
    "    dds <- dds[keep,]\n",
    "    cat(\"Percentiles after removing genes with sum(counts) <\", opt$min_count, \"\\n\")\n",
    "    quantile(rowSums(counts(dds)), probs = c(0, 0.1, 0.25, 0.5, 0.75, 0.9, 1))\n",
    "    # Get relevant percentiles\n",
    "    quant <- quantile(rowSums(counts(dds)), probs = c(0, 0.1, 0.25, 0.5, 0.75, 0.9, 1))\n",
    "    # Remove genes with <= 0.1 percentile counts\n",
    "    cat(\"Removing genes with sum(counts) <=\", quant[[2]])\n",
    "    keep <- rowSums(counts(dds)) > quant[[2]]\n",
    "    dds <- dds[keep,]\n",
    "    # Keep genes only where there are at least 3 samples having at least that count\n",
    "    cat(\"Removing genes with sum(counts) <=\", quant[[2]], \"in at least\", round(length(colnames(dds))/4), \"samples\\n\")\n",
    "    keep <- rowSums(counts(dds) > quant[[2]]) >= round(length(colnames(dds))/4)\n",
    "    dds <- dds[keep,]\n",
    "    \n",
    "    cat(\"Genes and samples after filtering raw counts:\\n\", dim(dds), \"\\n\")\n",
    "  }\n",
    "  \n",
    "  # ========== Normalize and Compute DEG Stats =============\n",
    "  # Create DESeq object\n",
    "  dds <- DESeq(dds)\n",
    "  \n",
    "  # Write filtered count matrix to file\n",
    "  write.table(counts(dds), file=paste(output_prefix, \"count_mtx_filtered.csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  # Write normalized values to file\n",
    "  normalized_count_mtx <- counts(dds, normalized=TRUE)\n",
    "  write.table(normalized_count_mtx, file=paste(output_prefix, \"count_mtx_normalized.csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  # Obtain DEG results and output to file\n",
    "  res <- results(dds, contrast=c(\"Condition\", combs[[c]][1], combs[[c]][2]))\n",
    "  res[['FoldChange']] <- 2^abs(res[['log2FoldChange']])*(res[['log2FoldChange']]/abs(res[['log2FoldChange']]))\n",
    "  write.table(res[order(res$log2FoldChange, decreasing=TRUE), ], file=paste(output_prefix, \"DESeq2_FullResult_\", comparison, \".csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  #resLFC <- lfcShrink(dds, coef=resultsNames(dds)[-1], type=\"apeglm\")\n",
    "  \n",
    "  # Filter out normalized low-expressed genes (bottom 10th percentile) by baseMean\n",
    "  cat(\"Genes and samples before filtering normalized genes:\\n\", dim(res), \"\\n\")\n",
    "  quant <- quantile(res$baseMean, probs = c(0, 0.1, 0.25, 0.5, 0.75, 0.9, 1))\n",
    "  cat(\"Removing genes with baseMean expression in bottom 10th percentile <=\", quant[[2]], \"\\n\")\n",
    "  res <- res[res$baseMean >= quant[[2]], ]\n",
    "  cat(\"Genes and samples after filtering normalized genes:\\n\", dim(res), \"\\n\")\n",
    "  \n",
    "  # Generate plots\n",
    "  png(paste(output_prefix, 'MAplot.png', sep=''))\n",
    "  plotMA(res)\n",
    "  abline(h=c((0-opt$lfc), opt$lfc), col=\"red\", lwd=2)\n",
    "  invisible(capture.output(dev.off()))\n",
    "  \n",
    "  #png(paste(output_prefix, 'resLFC_MAplot.png', sep=''))\n",
    "  #plotMA(resLFC, ylim=c(-2,2))\n",
    "  #invisible(capture.output(dev.off()))\n",
    "  \n",
    "  #png(paste(output_prefix, 'res_genecount_minpadj.png', sep=''))\n",
    "  #plotCounts(dds, gene=which.min(res$padj), intgroup=\"Condition\")\n",
    "  #invisible(capture.output(dev.off()))\n",
    "  \n",
    "  vsd <- vst(dds, blind=FALSE)\n",
    "  png(paste(output_prefix, 'PCAplot.png', sep=''))\n",
    "  plotPCA(vsd, intgroup=c(\"Condition\"))\n",
    "  invisible(capture.output(dev.off()))\n",
    "  \n",
    "  # ========== Organize DEGs ==============\n",
    "  # Remove any NAs\n",
    "  #res <- res[rowSums(is.na(res)) == 0, ]\n",
    "  \n",
    "  # Get significantly Upregulated (log2FoldChange > 0)\n",
    "  res_up <- res[res$log2FoldChange >= opt$lfc ,]\n",
    "  res_up_sorted <- res_up[order(res_up$log2FoldChange, decreasing=TRUE),]\n",
    "  write.table(res_up_sorted, file=paste(output_prefix, \"DESeq2_Result_\", combs[[c]][1], \".csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  # Get significantly Downregulated (log2FoldChange < 0)\n",
    "  res_down <- res[res$log2FoldChange <= (0 - opt$lfc) ,]\n",
    "  res_down_sorted <- res_down[order(res_down$log2FoldChange, decreasing=TRUE),]\n",
    "  write.table(res_down[order(res_down$log2FoldChange, decreasing=FALSE),], file=paste(output_prefix, \"DESeq2_Result_\", combs[[c]][2], \".csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  # Get all significantly changed genes\n",
    "  res_changed <- res[(res$log2FoldChange >= opt$lfc | res$log2FoldChange <= (0 - opt$lfc)),]\n",
    "  res_changed_sorted <- res_changed[order(abs(res_changed$log2FoldChange), decreasing=TRUE),]\n",
    "  write.table(res_changed_sorted, file=paste(output_prefix, \"DESeq2_Result_DEGs\", \".csv\", sep=\"\"), sep=\",\", quote=F, col.names=NA)\n",
    "  \n",
    "  genes <- list()\n",
    "  genes[[combs[[c]][1]]] <- rownames(res_up_sorted)\n",
    "  genes[[combs[[c]][2]]] <- rownames(res_down_sorted)\n",
    "  genes[[\"DEGs\"]] <- rownames(res_changed_sorted)\n",
    "  \n",
    "  genes_entrez <- list()\n",
    "  for (n in names(genes)){\n",
    "    if (tolower(opt$organism) == \"human\"){\n",
    "      genes_entrez[[n]] <- mapIds(org.Hs.eg.db, keys = genes[[n]], column = \"ENTREZID\", keytype = \"SYMBOL\")\n",
    "    }else if (tolower(opt$organism) == \"mouse\"){\n",
    "      genes_entrez[[n]] <- mapIds(org.Mm.eg.db, keys = genes[[n]], column = \"ENTREZID\", keytype = \"SYMBOL\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # ============= GO and KEGG Annotations ==============\n",
    "  for (n in names(genes)){\n",
    "    \n",
    "    # GO Annotation\n",
    "    tryCatch(\n",
    "      {\n",
    "        compGO <- compareCluster(geneCluster=genes[n],\n",
    "                                 OrgDb=annoDb,\n",
    "                                 fun=\"enrichGO\",\n",
    "                                 keyType=\"SYMBOL\", #ENSEMBL\n",
    "                                 pvalueCutoff=0.05,\n",
    "                                 pAdjustMethod=\"BH\",\n",
    "                                 readable=TRUE\n",
    "        ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "        if (!is.null(compGO)){\n",
    "          plt <- dotplot(compGO, showCategory = 8, title = paste(\"GO -\", n, sep=\"\"))\n",
    "          invisible(capture.output(ggsave(filename=paste(output_prefix, 'GO_annotation_', n, '.png', sep=''), plot=plt, dpi=320)))\n",
    "          # Write annotations to csv\n",
    "          write.table(as.data.frame(compGO), file=paste(output_prefix, 'GO_annotation_', n, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "        }\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "      }\n",
    "    )\n",
    "    \n",
    "    # KEGG Annotation\n",
    "    tryCatch(\n",
    "      {\n",
    "        # fun is \"groupGO\", \"enrichGO\", \"enrichKEGG\", \"enrichDO\" or \"enrichPathway\" \n",
    "        compKEGG <- compareCluster(geneCluster=genes_entrez[n],\n",
    "                                   fun=\"enrichKEGG\",\n",
    "                                   pvalueCutoff=0.05,\n",
    "                                   pAdjustMethod=\"BH\",\n",
    "                                   organism=keggOrg\n",
    "        ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "        if (!is.null(compKEGG)){\n",
    "          plt <- dotplot(compKEGG, showCategory = 8, title = paste(\"KEGG -\", n, sep=\"\"))\n",
    "          invisible(capture.output(ggsave(filename=paste(output_prefix, 'KEGG_annotation_', n, '.png', sep=''), plot=plt, dpi=320)))\n",
    "          \n",
    "          # Write annotations to csv\n",
    "          write.table(as.data.frame(compKEGG), file=paste(output_prefix, 'KEGG_annotation_', n, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "        }\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "      }\n",
    "    ) \n",
    "    \n",
    "    # GSEA\n",
    "    tryCatch(\n",
    "      {\n",
    "        original_gene_list <- res[genes[[n]],]$log2FoldChange\n",
    "        names(original_gene_list) <- rownames(res[genes[[n]],])\n",
    "        gene_list <- na.omit(original_gene_list)\n",
    "        \n",
    "        gsea <- gseGO(geneList=gene_list, \n",
    "                      ont =\"ALL\", \n",
    "                      keyType = \"SYMBOL\", \n",
    "                      nPerm = 10000, \n",
    "                      minGSSize = 3, \n",
    "                      maxGSSize = 800, \n",
    "                      pvalueCutoff = 0.05, \n",
    "                      verbose = TRUE, \n",
    "                      OrgDb = annoDb, \n",
    "                      pAdjustMethod = \"BH\"\n",
    "        )\n",
    "        if (!is.null(gsea)){\n",
    "          plt <- dotplot(gsea, showCategory = 8, title = paste(\"GSEA -\", n, sep=\"\"))\n",
    "          invisible(capture.output(ggsave(filename=paste(output_prefix, 'GSEA_annotation_', n, '.png', sep=''), plot=plt, dpi=320)))\n",
    "          \n",
    "          # Write annotations to csv\n",
    "          write.table(as.data.frame(compKEGG), file=paste(output_prefix, 'GSEA_annotation_', n, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "        }\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "      }\n",
    "    ) \n",
    "  }\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d367e0",
   "metadata": {},
   "source": [
    "## <center> =========== END ============ </center>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
