{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f10c1ed7",
   "metadata": {},
   "source": [
    "# <center> Differential Binding Analysis Notebook </center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa91750",
   "metadata": {},
   "source": [
    "## <center>Preface</center>  \n",
    "<p><center>This Notebook provides an interactive way to process, analyze, and view differential gene expressions from next-generation sequencing data, in particular from ChIP-Seq, CUTandTag, ATAC-Seq, etc....For RNA-Seq data, please refer to the RNA-Seq Differential Gene Expression Analysis Notebook<br>The following steps serve as a guide, however, care should be taken when executing and interpreting results.</center><br>The following R packages form the basis of tools used, but more may be required to run cells and are listed in the first step of loading packages:  \n",
    "    <li>  \n",
    "      <a href=\"https://bioconductor.org/packages/release/bioc/html/DiffBind.html\"><strong>DiffBind</strong></a> to compare peaksets (.bed files), determine overlaps, establish consensus peaksets, and count reads.<br>  \n",
    "      <a href=\"https://bioconductor.org/packages/release/bioc/html/DESeq2.html\"><strong>DESeq2</strong></a> to quantify differential binding statistics. <br>  \n",
    "      <a href=\"https://bioconductor.org/packages/release/bioc/html/edgeR.html\"><strong>edgeR</strong></a> to quantify differential binding statistics. <br>  \n",
    "      <a href=\"https://bioconductor.org/packages/release/bioc/html/ChIPseeker.html\"><strong>ChIPseeker</strong></a> to annotate binding sites.<br>  \n",
    "      <a href=\"https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html\"><strong>clusterProfiler</strong></a> to perform functional pathway analysis (GO/KEGG).<br>  \n",
    "      <p>Additional packages can be found in Renv_packages.txt</p>  \n",
    "    </li>\n",
    "    <center><br><em>Note that despite following the framework below, it is up to <strong>you</strong> to determine which steps, values, and parameters to change or ignore based on your knowledge of experimental details and data exploration to achieve your analysis goals.</em></center></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0563cf6",
   "metadata": {},
   "source": [
    "### sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ef8bc78",
   "metadata": {},
   "source": [
    "R version 4.3.1 (2023-06-16 ucrt)\n",
    "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
    "Running under: Windows 10 x64 (build 19045)\n",
    "\n",
    "Matrix products: default\n",
    "\n",
    "\n",
    "locale:\n",
    "[1] LC_COLLATE=English_Canada.utf8  LC_CTYPE=English_Canada.utf8    LC_MONETARY=English_Canada.utf8\n",
    "[4] LC_NUMERIC=C                    LC_TIME=English_Canada.utf8    \n",
    "\n",
    "time zone: America/Toronto\n",
    "tzcode source: internal\n",
    "\n",
    "attached base packages:\n",
    "[1] grid      stats4    stats     graphics  grDevices utils     datasets  methods   base     \n",
    "\n",
    "other attached packages:\n",
    " [1] optparse_1.7.5                          qvalue_2.32.0                          \n",
    " [3] reticulate_1.39.0                       pheatmap_1.0.12                        \n",
    " [5] RColorBrewer_1.1-3                      gridExtra_2.3                          \n",
    " [7] eulerr_7.0.2                            VennDiagram_1.7.3                      \n",
    " [9] futile.logger_1.4.3                     UpSetR_1.4.0                           \n",
    "[11] vulcan_1.22.0                           locfit_1.5-9.10                        \n",
    "[13] viper_1.34.0                            zoo_1.8-12                             \n",
    "[15] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 GenomicFeatures_1.52.2                 \n",
    "[17] AnnotationDbi_1.62.2                    ChIPpeakAnno_3.34.1                    \n",
    "[19] clusterProfiler_4.8.3                   ReactomePA_1.44.0                      \n",
    "[21] ChIPseeker_1.36.0                       lubridate_1.9.3                        \n",
    "[23] forcats_1.0.0                           stringr_1.5.1                          \n",
    "[25] dplyr_1.1.4                             purrr_1.0.2                            \n",
    "[27] readr_2.1.5                             tidyr_1.3.1                            \n",
    "[29] tibble_3.2.1                            ggplot2_3.5.1                          \n",
    "[31] tidyverse_2.0.0                         DiffBind_3.10.1                        \n",
    "[33] SummarizedExperiment_1.30.2             Biobase_2.60.0                         \n",
    "[35] MatrixGenerics_1.12.3                   matrixStats_1.4.1                      \n",
    "[37] GenomicRanges_1.52.0                    GenomeInfoDb_1.36.4                    \n",
    "[39] IRanges_2.34.1                          S4Vectors_0.38.1                       \n",
    "[41] BiocGenerics_0.46.0                    \n",
    "\n",
    "loaded via a namespace (and not attached):\n",
    "  [1] segmented_2.1-2          fs_1.6.4                 bitops_1.0-8             enrichplot_1.20.3       \n",
    "  [5] HDO.db_0.99.1            httr_1.4.7               InteractionSet_1.28.1    numDeriv_2016.8-1.1     \n",
    "  [9] tools_4.3.1              utf8_1.2.4               R6_2.5.1                 lazyeval_0.2.2          \n",
    " [13] apeglm_1.22.1            withr_3.0.1              graphite_1.46.0          prettyunits_1.2.0       \n",
    " [17] cli_3.6.1                formatR_1.14             scatterpie_0.2.4         SQUAREM_2021.1          \n",
    " [21] mvtnorm_1.3-1            proxy_0.4-27             mixsqp_0.3-54            Rsamtools_2.16.0        \n",
    " [25] yulab.utils_0.1.7        gson_0.1.0               DOSE_3.26.2              plotrix_3.8-4           \n",
    " [29] BSgenome_1.68.0          invgamma_1.1             bbmle_1.0.25.1           limma_3.56.2            \n",
    " [33] rstudioapi_0.16.0        RSQLite_2.3.7            generics_0.1.3           gridGraphics_0.5-1      \n",
    " [37] BiocIO_1.10.0            hwriter_1.3.2.1          gtools_3.9.5             GO.db_3.17.0            \n",
    " [41] Matrix_1.5-4.1           interp_1.1-6             fansi_1.0.6              lifecycle_1.0.4         \n",
    " [45] edgeR_3.42.4             yaml_2.3.10              gplots_3.2.0             BiocFileCache_2.8.0     \n",
    " [49] paletteer_1.6.0          blob_1.2.4               crayon_1.5.3             bdsmatrix_1.3-7         \n",
    " [53] lattice_0.21-8           cowplot_1.1.3            KEGGREST_1.40.1          metapod_1.8.0           \n",
    " [57] pillar_1.9.0             fgsea_1.26.0             rjson_0.2.23             boot_1.3-28.1           \n",
    " [61] systemPipeR_2.6.3        codetools_0.2-19         fastmatch_1.1-4          glue_1.7.0              \n",
    " [65] ShortRead_1.58.0         downloader_0.4           ggfun_0.1.6              GreyListChIP_1.32.1     \n",
    " [69] data.table_1.16.0        vctrs_0.6.5              png_0.1-8                treeio_1.24.3           \n",
    " [73] gtable_0.3.5             kernlab_0.9-33           rematch2_2.1.2           amap_0.8-19.1           \n",
    " [77] emdbook_1.3.13           cachem_1.1.0             S4Arrays_1.0.4           tidygraph_1.3.1         \n",
    " [81] coda_0.19-4.1            survival_3.5-5           nlme_3.1-162             ggtree_3.8.2            \n",
    " [85] bit64_4.5.2              progress_1.2.3           filelock_1.0.3           irlba_2.3.5.1           \n",
    " [89] KernSmooth_2.23-21       colorspace_2.1-1         DBI_1.2.3                DESeq2_1.40.2           \n",
    " [93] tidyselect_1.2.1         bit_4.5.0                compiler_4.3.1           curl_5.2.3              \n",
    " [97] graph_1.78.0             csaw_1.34.0              xml2_1.3.6               plotly_4.10.4           \n",
    "[101] DelayedArray_0.26.6      shadowtext_0.1.4         rtracklayer_1.60.1       scales_1.3.0            \n",
    "[105] caTools_1.18.3           RBGL_1.76.0              rappdirs_0.3.3           digest_0.6.37           \n",
    "[109] mixtools_2.0.0           XVector_0.40.0           htmltools_0.5.8.1        pkgconfig_2.0.3         \n",
    "[113] jpeg_0.1-10              dbplyr_2.5.0             regioneR_1.32.0          fastmap_1.2.0           \n",
    "[117] rlang_1.1.4              htmlwidgets_1.6.4        farver_2.1.2             jsonlite_1.8.9          \n",
    "[121] BiocParallel_1.34.2      GOSemSim_2.26.1          RCurl_1.98-1.16          magrittr_2.0.3          \n",
    "[125] GenomeInfoDbData_1.2.10  ggplotify_0.1.2          wordcloud_2.6            patchwork_1.3.0         \n",
    "[129] munsell_0.5.1            Rcpp_1.0.13              ape_5.8                  viridis_0.6.5           \n",
    "[133] stringi_1.8.4            ggraph_2.2.1             zlibbioc_1.46.0          MASS_7.3-60             \n",
    "[137] plyr_1.8.9               parallel_4.3.1           ggrepel_0.9.6            deldir_2.0-4            \n",
    "[141] Biostrings_2.68.1        graphlayouts_1.2.0       splines_4.3.1            multtest_2.56.0         \n",
    "[145] hms_1.1.3                igraph_2.0.3             reshape2_1.4.4           biomaRt_2.56.1          \n",
    "[149] futile.options_1.0.1     XML_3.99-0.17            latticeExtra_0.6-30      lambda.r_1.2.4          \n",
    "[153] tzdb_0.4.0               tweenr_2.0.3             getopt_1.20.4            polyclip_1.10-7         \n",
    "[157] ashr_2.2-63              ggforce_0.4.2            restfulr_0.0.15          reactome.db_1.84.0      \n",
    "[161] e1071_1.7-16             tidytree_0.4.6           class_7.3-22             viridisLite_0.4.2       \n",
    "[165] truncnorm_1.0-9          aplot_0.2.3              memoise_2.0.1            GenomicAlignments_1.36.0\n",
    "[169] timechange_0.3.0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f24f257",
   "metadata": {},
   "source": [
    "### Required:\n",
    "<br><strong>1) A <u>DIFFBIND SAMPLESHEET</u> manually created by you as a comma-separated file (.csv) to indicate sample groups and file info, example formatted as follows:</strong><br>\n",
    "<table style=’float:left;’>\n",
    "    <tr>\n",
    "        <th>SampleID</th>\n",
    "        <th>Condition</th>\n",
    "        <th>Replicate</th>\n",
    "        <th>Peaks</th>\n",
    "        <th>PeakCaller</th>\n",
    "        <th>PeakFormat</th>\n",
    "        <th>bamReads</th>\n",
    "        <th>Factor</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>sample1-1</td>\n",
    "        <td>control</td>\n",
    "        <td>1</td>\n",
    "        <td>sample1-1_peaks.narrowPeak</td>\n",
    "        <td>bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>sample1-1.bam</td>\n",
    "        <td>MACS</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>sample1-2</td>\n",
    "        <td>control</td>\n",
    "        <td>2</td>\n",
    "        <td>sample1-2_peaks.bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>sample1-2.bam</td>\n",
    "        <td>SEACR</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>sample2-1</td>\n",
    "        <td>experiment</td>\n",
    "        <td>1</td>\n",
    "        <td>sample2-1_peaks.narrowPeak</td>\n",
    "        <td>bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>sample2-1.bam</td>\n",
    "        <td>MACS</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td>sample2-2</td>\n",
    "        <td>experiment</td>\n",
    "        <td>2</td>\n",
    "        <td>sample2-2_peaks.bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>bed</td>\n",
    "        <td>sample2-2.bam</td>\n",
    "        <td>SEACR</td>\n",
    "    </tr>\n",
    "</table><br>\n",
    "&emsp;<strong>SampleID must be unique for all rows.</strong><br>\n",
    "&emsp;<strong>At least 2 replicates in each sample condition are required for DiffBind affinity analysis.</strong><br>\n",
    "&emsp;<strong>Be sure to put the full path for the peaks and bam files.</strong><br>\n",
    "&emsp;<strong>Factor column may be omitted if one peak caller was used for .bed files.</strong><br>\n",
    "<center><strong>---This file must be produced manually prior to running the complete script.---</strong></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2964963f",
   "metadata": {},
   "source": [
    "## <center> ======= Load Packages ======= </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e76db7f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "suppressWarnings(suppressPackageStartupMessages({\n",
    "  library(DiffBind)\n",
    "  library(tidyverse)\n",
    "  library(ChIPseeker)\n",
    "  library(ReactomePA)\n",
    "  library(clusterProfiler)\n",
    "  library(vulcan)\n",
    "  library(UpSetR)\n",
    "  library(dplyr)\n",
    "  library(VennDiagram)\n",
    "  library(eulerr)\n",
    "  require(gridExtra)\n",
    "  library(RColorBrewer)\n",
    "  library(pheatmap)\n",
    "  library(grid)\n",
    "  library(reticulate)\n",
    "  library(qvalue)\n",
    "  library(optparse)\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "483fb1a6",
   "metadata": {},
   "source": [
    "## <center> =========== Define Functions ============ </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a6e618",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Load annotation database according to input params\n",
    "load_annotation <- function(assembly, database){\n",
    "  # ========= Get database references for annotations =========\n",
    "  if (assembly == \"mm10\" | assembly == \"mm9\"){\n",
    "    library(org.Mm.eg.db)\n",
    "    annoDb <- \"org.Mm.eg.db\"\n",
    "    keggOrg <- \"mmu\"\n",
    "    if (database == \"ucsc\"){\n",
    "      if (assembly == \"mm10\"){\n",
    "        library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
    "        txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene\n",
    "      }else if (assembly == \"mm9\"){\n",
    "        library(TxDb.Mmusculus.UCSC.mm9.knownGene)\n",
    "        txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene\n",
    "      }\n",
    "    }else if (database == \"ensemble\"){\n",
    "      library(EnsDb.Mmusculus.v79)\n",
    "      txdb <- EnsDb.Mmusculus.v79\n",
    "      seqlevelsStyle(txdb) <- \"UCSC\" # format ensembl genes using UCSC style\n",
    "    }\n",
    "    else{\n",
    "      stop(\"Invalid choice of annotation database\")\n",
    "    }\n",
    "  } else if (assembly == \"hg38\" | assembly == \"hg19\"){\n",
    "    library(org.Hs.eg.db)\n",
    "    annoDb <- \"org.Hs.eg.db\"\n",
    "    keggOrg <- \"hsa\"\n",
    "    if (database == \"ucsc\"){\n",
    "      if (assembly == \"hg38\"){\n",
    "        library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "        txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "      }else if (assembly == \"hg19\"){\n",
    "        library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "        txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene\n",
    "      }\n",
    "    }else if (database == \"ensemble\"){\n",
    "      library(EnsDb.Hsapiens.v86)\n",
    "      txdb <- EnsDb.Hsapiens.v86\n",
    "      seqlevelsStyle(txdb) <- \"UCSC\" # format ensembl genes using UCSC style\n",
    "    } \n",
    "    else{\n",
    "      stop(\"Invalid choice of annotation database\")\n",
    "    }\n",
    "  } else if (assembly == \"rn6\"){\n",
    "    library(TxDb.Rnorvegicus.UCSC.rn6.refGene)\n",
    "    library(org.Rn.eg.db)\n",
    "    annoDb <- \"org.Rn.eg.db\"\n",
    "    keggOrg <- \"rno\"\n",
    "    if (database == \"ucsc\"){\n",
    "      txdb <- TxDb.Rnorvegicus.UCSC.rn6.refGene\n",
    "    }else if (database == \"ensemble\"){\n",
    "      library(EnsDb.Rnorvegicus.v79)\n",
    "      txdb <- EnsDb.Rnorvegicus.v79\n",
    "      seqlevelsStyle(txdb) <- \"UCSC\" # format ensembl genes using UCSC style\n",
    "    } \n",
    "    else{\n",
    "      stop(\"Invalid choice of annotation database\")\n",
    "    }\n",
    "  } else{\n",
    "    stop(\"Invalid choice of assembly\")\n",
    "  }\n",
    "  \n",
    "  anno_ref <- list(\"txdb\"=txdb, \"annoDb\"=annoDb, \"keggOrg\"=keggOrg)\n",
    "  \n",
    "  return(anno_ref)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05b2c57",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Dotplots for GO/KEGG pathway analysis\n",
    "make_dotplot <- function(df, title=\"\", ylabel=\"Description\", colour=\"#56B1F7\", n=15){\n",
    "  df$ycolour <- \"black\"\n",
    "  if (\"ONTOLOGY\" %in% colnames(df)){\n",
    "    df$Description <- paste(df$ONTOLOGY, df$Description, sep=' - ')\n",
    "    df$ycolour <- ifelse(grepl(\"BP -\", df$Description), 'blue', df$ycolour)\n",
    "    df$ycolour <- ifelse(grepl(\"CC -\", df$Description), 'red', df$ycolour)\n",
    "    df$ycolour <- ifelse(grepl(\"MF -\", df$Description), 'darkgreen', df$ycolour)\n",
    "  }\n",
    "  df <- df[order(df$p.adjust, decreasing=FALSE),]\n",
    "  \n",
    "  # Re-format y-axis labels to not squish graph\n",
    "  for (d in 1:length(df$Description)){\n",
    "    i <- 1\n",
    "    s <- df$Description[d]\n",
    "    s <- str_remove(s, \" - Mus musculus \\\\(house mouse\\\\)\")\n",
    "    df$Description[d] <- \"\"\n",
    "    while (i < length(strsplit(s, ' ')[[1]]) + 1){\n",
    "      df$Description[d] <- paste(df$Description[d], paste(strsplit(s, ' ')[[1]][i:(i+5)], collapse = ' '), sep='\\n')\n",
    "      i <- (i+6)\n",
    "    }\n",
    "    df$Description[d] <- gsub(\" NA\", \"\", df$Description[d])\n",
    "    df$Description[d] <- substring(df$Description[d], 2, nchar(df$Description[d]))\n",
    "  }\n",
    "  \n",
    "  # Plot\n",
    "  plt <- ggplot() +\n",
    "    geom_point(data=head(df, n=n),\n",
    "               aes(x = -log(p.adjust), \n",
    "                   y = reorder(Description, -p.adjust), \n",
    "                   colour = Count, \n",
    "                   size = unname(unlist(sapply(GeneRatio, function(x) eval(parse(text=x)))))*100,\n",
    "               ),\n",
    "    ) + \n",
    "    theme_classic() +\n",
    "    theme(axis.text.y = element_text(colour=rev(head(df$ycolour, n=n)))) +\n",
    "    scale_color_gradient(low = \"black\", high = colour) +\n",
    "    ggtitle(title) \n",
    "  plt$labels$x <- \"-log(p.adjust)\"\n",
    "  plt$labels$y <- ylabel\n",
    "  plt$labels$size <- \"GenePercentage\"\n",
    "  plt$labels$colour <- \"GeneCount\"\n",
    "  return(plt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0a7cd04",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "make_pheatmapplot <- function(anno, res, anno_type=\"GO\", assembly='mm10', heat_colour=\"PiYG\", num_terms=25, num_genes=50, lfc=0.6, dendro=TRUE, sort_genes=TRUE, title=\"\", xlabel=\"Gene\", ylabel=\"Term\"){\n",
    "  # colour should be \"Reds\", \"Greens\", \"Blues\", or \"PiYG\"\n",
    "  if (\"ONTOLOGY\" %in% colnames(anno)){\n",
    "    anno$Description <- paste(anno$ONTOLOGY, anno$Description, sep=' - ')\n",
    "  }\n",
    "  \n",
    "  # Take top n terms (most significant, already sorted by padj)\n",
    "  df <- head(anno[order(anno$p.adjust, decreasing=FALSE), ], n=num_terms)\n",
    "  \n",
    "  # Re-format y-axis labels to not squish graph\n",
    "  for (d in 1:length(df$Description)){\n",
    "    i <- 1\n",
    "    s <- df$Description[d]\n",
    "    df$Description[d] <- \"\"\n",
    "    while (i < length(strsplit(s, ' ')[[1]])){\n",
    "      df$Description[d] <- paste(df$Description[d], paste(strsplit(s, ' ')[[1]][i:(i+5)], collapse = ' '), sep='\\n')\n",
    "      i <- (i+6)\n",
    "    }\n",
    "    df$Description[d] <- gsub(\" NA\", \"\", df$Description[d])\n",
    "    df$Description[d] <- substring(df$Description[d], 2, nchar(df$Description[d]))\n",
    "  }\n",
    "  \n",
    "  # Create dataframe (matrix) of annotation terms vs genes with gene's associated log2FoldChange\n",
    "  d <- data.frame()\n",
    "  for (a in df$Description){\n",
    "    gene_group <- strsplit(df[df$Description == a, ]$geneID, '/')[[1]]\n",
    "    # For KEGG to convert EntrezID to gene Symbol\n",
    "    if (anno_type == \"KEGG\"){\n",
    "      if (assembly == \"hg19\" | assembly == \"hg38\"){\n",
    "        gene_group <- mapIds(org.Hs.eg.db, keys = gene_group, column = \"SYMBOL\", keytype = \"ENTREZID\")\n",
    "      }else if (assembly == \"mm10\" | assembly == \"mm9\"){\n",
    "        gene_group <- mapIds(org.Mm.eg.db, keys = gene_group, column = \"SYMBOL\", keytype = \"ENTREZID\")\n",
    "      }else if (assembly == \"rn6\"){\n",
    "        gene_group <- mapIds(org.Rn.eg.db, keys = gene_group, column = \"SYMBOL\", keytype = \"ENTREZID\")\n",
    "      }\n",
    "    }\n",
    "    temp <- res[res$SYMBOL %in% gene_group, ]\n",
    "    temp <- aggregate(temp, list(temp$SYMBOL), mean)\n",
    "    row.names(temp) <- temp$Group.1\n",
    "    d[temp$Group.1, a] <- temp$log2FoldChange\n",
    "  }\n",
    "  # Sort by genes instead of by term (i.e. number of times gene found in all top terms)\n",
    "  if (sort_genes){\n",
    "    d <- d[names(sort(rowSums(is.na(d)))), ]\n",
    "    xlabel <- paste(xlabel, \"(Most Frequent in Top Terms)\")\n",
    "  }\n",
    "  # Set NA to 0\n",
    "  d[is.na(d)] <- 0\n",
    "  \n",
    "  range_max <- round(max(apply(d, 2, max)))\n",
    "  range_min <- round(min(apply(d, 2, min)))\n",
    "  #breaks <- seq( -round(2^lfc), round(2^lfc), length.out = 101)\n",
    "  breaks <- seq( -abs((max(range_min, -round(lfc)))), min(range_max, round(lfc)), length.out = 101)\n",
    "  #breaks <- seq( -abs((max(range_min, -round(2^lfc)))), min(range_max, round(2^lfc)), length.out = 101)\n",
    "  if (range_max == 0){\n",
    "    color <- rev(colorRampPalette(brewer.pal(n = 11, name = heat_colour))(101))\n",
    "  }else{\n",
    "    color <- colorRampPalette(brewer.pal(n = 11, name = heat_colour))(101)\n",
    "  }\n",
    "  \n",
    "  # Plot\n",
    "  setHook(\"grid.newpage\", function() pushViewport(viewport(x=0,y=0.05,width=0.95, height=0.95, name=\"vp\", just=c(\"left\",\"bottom\"))), action=\"prepend\")\n",
    "  pheatmap(t(head(d, n=num_genes)), \n",
    "           border_color = \"grey90\",\n",
    "           color = color, # \"Reds, Greens, Blues, RdYlGn for DEGs\n",
    "           fontsize_row = 5,\n",
    "           fontsize_col = 5,\n",
    "           na_col = \"white\",\n",
    "           breaks = breaks,\n",
    "           cluster_rows = dendro,\n",
    "           cluster_cols = dendro,\n",
    "           main = title,\n",
    "  ) \n",
    "  setHook(\"grid.newpage\", NULL, \"replace\")\n",
    "  grid.text(\"log2FoldChange\", x=0.95, y=0.875, gp=gpar(fontsize=8))\n",
    "  grid.text(xlabel, y=0, gp=gpar(fontsize=14))\n",
    "  grid.text(ylabel, x=1, y=0.35,  rot=270, gp=gpar(fontsize=14))\n",
    "  plt <- grid.grab()\n",
    "  \n",
    "  return(plt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08a154f6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# For easily changing directories for outputs\n",
    "change_dirs <- function(res_dir, method, subfolder){\n",
    "  if ((method == DBA_DESEQ2) | (method == 'DESeq2')){\n",
    "    method_dir <- paste(res_dir, 'DESeq2/', sep='')\n",
    "    if (!file.exists(paste(method_dir, subfolder, '/', sep=''))) {\n",
    "      dir.create(paste(method_dir, subfolder, '/', sep=''))\n",
    "    }\n",
    "  }else if ((method == DBA_EDGER) | (method == 'edgeR')){\n",
    "    method_dir <- paste(res_dir, 'edgeR/', sep='')\n",
    "    if (!file.exists(paste(method_dir, subfolder, '/', sep=''))) {\n",
    "      dir.create(paste(method_dir, subfolder, '/', sep=''))\n",
    "    }\n",
    "  }else{\n",
    "    method_dir <- paste(res_dir, method, '/', sep='')\n",
    "    if (!file.exists(paste(method_dir, subfolder, '/', sep=''))) {\n",
    "      dir.create(paste(method_dir, subfolder, '/', sep=''))\n",
    "    }\n",
    "  }\n",
    "  return(paste(method_dir, subfolder, '/', sep=''))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8a525e",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf056baf",
   "metadata": {},
   "source": [
    "## <center> ======= Define Input Parameters ======= </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f31e366",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ======= Get command-line optional arguments (defaults) =======\n",
    "option_list = list(\n",
    "  make_option(c(\"-samplesheet\", \"--samplesheet\"), type=\"character\", default=NULL, help=\"DiffBind-formatted sample sheet\", metavar=\"character\"),\n",
    "  make_option(c(\"-a\", \"--assembly\"), type=\"character\", default=\"mm10\", help=\"Assembly to annotate genes/peaks (e.g. hg19, hg38, mm9, mm10, rn6)\", metavar=\"character\"),\n",
    "  make_option(c(\"-r\", \"--result_dir\"), type=\"character\", default=\"Peaks_Analysis/\", help=\"Directory name for saving output results\", metavar=\"character\"),\n",
    "  make_option(c(\"-d\", \"--database\"), type=\"character\", default=\"ucsc\", help=\"Database reference for peaks gene annotations, ucsc (default) or ensembl\", metavar=\"character\"),\n",
    "  make_option(c(\"-l\", \"--annotation_level\"), type=\"character\", default=\"transcript\", help=\"Level parameter for annotatePeak, 'gene' or 'transcript'\", metavar=\"character\"),\n",
    "  make_option(c(\"--combine_callers\"), type=\"logical\", action=\"store_true\", default=FALSE, help=\"Flag to add peaks from callers instead of taking consensus peaks\", metavar=\"logical\"),\n",
    "  make_option(c(\"--combine_replicates\"), type=\"logical\", action=\"store_true\", default=FALSE, help=\"Flag to add peaks from replicates instead of taking consensus peaks\", metavar=\"logical\"),\n",
    "  make_option(c(\"-b\", \"--blacklisted_keep\"), type=\"logical\", action=\"store_true\", default=FALSE, help=\"Flag to keep blacklisted regions in raw peaks files\", metavar=\"logical\"),\n",
    "  make_option(c(\"--lfc\"), type=\"double\", default=0.585, help=\"Magnitude of log2foldchange to define significant up/down regulation of genes\", metavar=\"double\"),\n",
    "  make_option(c(\"--fdr\"), type=\"double\", default=0.05, help=\"Significance threshold (false discovery rate, a.k.a. p.adjust value) for DEGs\", metavar=\"double\"),\n",
    "  make_option(c(\"--occupancy_only\"), type=\"logical\", action=\"store_true\", default=FALSE, help=\"Flag to only perform peaks occupancy analysis\", metavar=\"logical\"),\n",
    "  make_option(c(\"--david_user\"), type=\"character\", default=\"earezza@ohri.ca\", help=\"User email for DAVID web tools (must be registered, https://david.ncifcrf.gov/content.jsp?file=DAVID_WebService.html)\", metavar=\"character\"),\n",
    "  make_option(c(\"--minGSSize\"), type=\"integer\", default=10, help=\"minimal size of genes annotated for testing\", metavar=\"integer\"),\n",
    "  make_option(c(\"--maxGSSize\"), type=\"integer\", default=500, help=\"maximal size of genes annotated for testing\", metavar=\"integer\")\n",
    ");\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30a55f77",
   "metadata": {},
   "source": [
    "#### Manually set parameters (if different from defaults)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85d94823",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Put samplesheet file path and directory path name ending in '/' for saving results\n",
    "opt$samplesheet <- \"\"\n",
    "opt$result_dir <- \"\"\n",
    "opt$assembly <- \"mm10\"\n",
    "opt$occupancy_only <- TRUE # skip affinity analysis (set to FALSE to perform complete analysis)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0974b73e",
   "metadata": {},
   "source": [
    "#### View current parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a221e6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# View params\n",
    "for (i in which(names(opt) != \"help\")) {\n",
    "  cat(names(opt)[i], '=', paste(opt)[i], \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3106ff3",
   "metadata": {},
   "source": [
    "#### Create output directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8404269",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Adjust dir string to end with '/'\n",
    "if (str_sub(opt$result_dir, -1) != \"/\"){\n",
    "  opt$result_dir = cat(opt$result_dir, '/', sep='')\n",
    "}\n",
    "setwd(dirname(opt$samplesheet))\n",
    "samplesheet <- basename(opt$samplesheet)\n",
    "\n",
    "# Create and set output directories\n",
    "if (!file.exists(opt$result_dir)) {\n",
    "  dir.create(opt$result_dir)\n",
    "}\n",
    "cat(\"Output files will be in\", opt$result_dir, \"\\n\")\n",
    "\n",
    "result_dir <- paste(opt$result_dir, \"Occupancy_Analysis/\", sep='')\n",
    "if (!file.exists(result_dir)) {\n",
    "  dir.create(result_dir)\n",
    "}\n",
    "\n",
    "supplementary_dir <- paste(result_dir, \"Supplementary/\", sep='')\n",
    "if (!file.exists(supplementary_dir)) {\n",
    "  dir.create(supplementary_dir)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93916c94",
   "metadata": {},
   "source": [
    "# <center> ========= START OCCUPANCY ANALYSIS ========= </center>\n",
    "In this <em>occupancy analysis</em> step, binding sites declared by peak caller(s) are used to identify differential and shared sites between conditions. Overlapping peaks between replicates/conditions are determined by the range of the peaks, where an overlap of 1bp is considered. In this case, using occupancy alone for differential binding site analysis provides a general and simplistic approach by only considering where peaks exist, not the extent of enrichment at those called peaks.  \n",
    "Following this, the <em>affinity analysis</em> step can provide a more rigorous statistical result where the read counts at binding sites are taken into account."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b9a8166",
   "metadata": {},
   "source": [
    "#### Load Input Files and Annotation Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "846f7710",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Load peaksets =========\n",
    "setwd(dirname(opt$samplesheet))\n",
    "samplesheet <- basename(opt$samplesheet)\n",
    "\n",
    "# Get annotations reference and respective promoter regions\n",
    "anno_ref <- load_annotation(opt$assembly, opt$database)\n",
    "promoters <- getPromoters(TxDb=anno_ref$txdb, upstream=3000, downstream=3000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c74731",
   "metadata": {},
   "source": [
    "Compute average fragment sizes (or manually set your expected size(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f099a4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Average for each sample\n",
    "fragment_size <- 1:length(read.csv(samplesheet)$SampleID)\n",
    "frag_sizes <- list()\n",
    "for (b in unique(read.csv(samplesheet)$Condition)){\n",
    "  for (r in unique(read.csv(samplesheet)$Replicate)){\n",
    "    png(paste(supplementary_dir, 'fragment_length_', b, '-', r, '.png', sep=\"\"))\n",
    "    mean_fragment_size <- average_fragment_length(read.csv(samplesheet)$bamReads[[which(read.csv(samplesheet)$Condition == b)[1]]], plot=TRUE)\n",
    "    for (i in which(read.csv(samplesheet)$Condition == b & read.csv(samplesheet)$Replicate == r)){\n",
    "      fragment_size[i] <- mean_fragment_size \n",
    "      frag_sizes[[b]] <- mean_fragment_size\n",
    "    }\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }\n",
    "}\n",
    "invisible(capture.output(gc())) \n",
    "#fragment_size <- 125 # default, sets the same for all samples\n",
    "#fragment_size <- c(125, 100, 100, 125) # if each sample is different, ordered by SampleID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4375b4c1",
   "metadata": {},
   "source": [
    "#### Create DiffBind object from samplesheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdcfc471",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dbObj <- dba(sampleSheet=samplesheet, \n",
    "            minOverlap=1,   # proportion of peaksets required that overlapping peaks will be included (1 = use all peaks, 0.5 = peaks must be in at least half of peaksets)\n",
    "            config=data.frame(th=opt$fdr,\n",
    "                #DataType=DBA_DATA_GRANGES, \n",
    "                RunParallel=TRUE,\n",
    "                minQCth=15, \n",
    "                fragmentSize=fragment_size, # Reads will be extended to this length before counting overlaps in affinity analysis\n",
    "                reportInit=\"DBA\",\n",
    "                bUsePval=FALSE\n",
    "                )\n",
    ")\n",
    "cat(\"Raw peaksets:\\n\")\n",
    "dbObj"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69385fa",
   "metadata": {},
   "source": [
    "Define colors and color code for figures (groups colors by Condition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9426556b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Colour codes for consistency in plots (add more colours if needed)\n",
    "colours <- c(\"#00BFC4\", \"#F8766D\", \"#7CAE00\", \"#C77CFF\", \"#e69e02\", \"#00A9FF\", \"#C77CFF\", \"#FF61CC\", \n",
    "             \"#FF0000\", \"#FF4D00\", \"#80FF00\", \"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\" , \"#D55E00\" , \"#CC79A7\"\n",
    ")\n",
    "conditions_colour_code <- list()\n",
    "for (i in 1:length(unique(dbObj$samples$Condition))) {\n",
    "  conditions_colour_code[[unique(dbObj$samples$Condition)[i]]] <- colours[i]\n",
    "}\n",
    "conditions_colour_code[['Shared']] <- \"grey\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "935076a6",
   "metadata": {},
   "source": [
    "### Plots for raw peaksets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d0f8c3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Heatmap correlation between samples\n",
    "png(filename=paste(supplementary_dir, 'raw_heatmap.png', sep=''))\n",
    "dba.plotHeatmap(dbObj)\n",
    "invisible(capture.output(dev.off()))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6216efb1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Show overlap rates for each condition\n",
    "cat(\"Peak overlaps in at least (1, 2, ...) replicates/callers for each condition:\\n\")\n",
    "png(paste(supplementary_dir, \"raw_overlap_rates.png\", sep=\"\"))\n",
    "par(mfrow=c(length(unique(dbObj$samples$Condition)), 1), mar = c(2, 4, 4, 2))\n",
    "for (c in unique(dbObj$samples$Condition)) {\n",
    "  cat('\\n', c, '\\n')\n",
    "  olap.rate <- dba.overlap(dbObj, dba.mask(dbObj, attribute=DBA_CONDITION, value=c, combine='or'), mode=DBA_OLAP_RATE)\n",
    "  cat(olap.rate)\n",
    "  cat('\\n')\n",
    "  plot(olap.rate, type='l', xlim=c(0.75, length(olap.rate)), ylab='# overlapping peaks', xlab='# peaksets (replicates and peak callers)', col=conditions_colour_code[[c]])\n",
    "  text(x=1:length(olap.rate), y=olap.rate, olap.rate, cex=0.75)\n",
    "  axis(side=1, at=1:length(olap.rate))\n",
    "  title(main=c(\"Overlap Rate\", c))\n",
    "}\n",
    "invisible(capture.output(dev.off()))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be832487",
   "metadata": {},
   "source": [
    "### Remove Blacklisted Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "614ab761",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Remove blacklisted regions to ignore irrelevant peaks (blacklisted regions from ENCODE, genome selected is based on prediction from bam files)\n",
    "tryCatch (\n",
    "  {\n",
    "    if (!opt$blacklisted_keep){\n",
    "      dbObj.noblacklist <- dba.blacklist(dbObj, blacklist=TRUE, greylist=FALSE)\n",
    "      blacklisted_peaks <- dba.blacklist(dbObj.noblacklist, Retrieve=DBA_BLACKLISTED_PEAKS)\n",
    "      cat(\"After blacklisted regions removed:\\n\")\n",
    "      dbObj.noblacklist\n",
    "    }else{\n",
    "      cat(\"\\nBlacklisted regions not removed, proceeding with raw peaksets...\\n\")\n",
    "      dbObj.noblacklist <- dbObj\n",
    "    }\n",
    "  },error = function(e)\n",
    "  {\n",
    "    message(e)\n",
    "  }\n",
    ")\n",
    "if (!exists(\"dbObj.noblacklist\")) {\n",
    "  cat(\"\\nBlacklisted regions not removed, proceeding with raw peaksets...\\n\")\n",
    "  dbObj.noblacklist <- dbObj\n",
    "  dbObj.noblacklist\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b41dfcc",
   "metadata": {},
   "source": [
    "### Plots for blacklist-removed peaksets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dacd32a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Heatmap correlation between samples with blacklisted regions removed\n",
    "png(paste(supplementary_dir, 'raw_noblacklist_heatmap.png', sep=''))\n",
    "dba.plotHeatmap(dbObj.noblacklist)\n",
    "invisible(capture.output( dev.off() ))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f8c5d08",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Show overlap rates for each condition\n",
    "cat(\"Peak overlaps in at least (1, 2, ...) replicates/callers for each condition:\\n\")\n",
    "png(paste(supplementary_dir, \"raw_noblacklist_overlap_rates.png\", sep=\"\"))\n",
    "par(mfrow=c(length(unique(dbObj$samples$Condition)), 1), mar = c(2, 4, 4, 2))\n",
    "for (c in unique(dbObj$samples$Condition)) {\n",
    "  cat('\\n', c, '\\n')\n",
    "  olap.rate <- dba.overlap(dbObj, dba.mask(dbObj.noblacklist, attribute=DBA_CONDITION, value=c, combine='or'), mode=DBA_OLAP_RATE)\n",
    "  cat(olap.rate)\n",
    "  cat('\\n')\n",
    "  plot(olap.rate, type='l', xlim=c(0.75, length(olap.rate)), ylab='# overlapping peaks', xlab='# peaksets (replicates and peak callers)', col=conditions_colour_code[[c]])\n",
    "  text(x=1:length(olap.rate), y=olap.rate, olap.rate, cex=0.75)\n",
    "  axis(side=1, at=1:length(olap.rate))\n",
    "  title(main=c(\"Overlap Rate\", c))\n",
    "}\n",
    "invisible(capture.output(dev.off()))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5410d6b",
   "metadata": {},
   "source": [
    "### Obtain consensus peaksets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7da800",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Get Consensus Peaks =========\n",
    "if (opt$combine_replicates == TRUE){\n",
    "  rep_overlaps <- 1 # add all peaks from replicates\n",
    "}else{\n",
    "  rep_overlaps <- 2 # add only overlapping peaks from replicates (peaks must be in at least 2 replicates) into consensus peakset\n",
    "}\n",
    "\n",
    "# If using peaks from multiple peak callers (defined in Factor column of samplesheet)\n",
    "if (length(unique(dbObj$samples$Factor)) > 1){\n",
    "  # Get consensus overlaps in peaksets for each condition (peaks must be overlapping in majority (2/3rds) of peak callers) or add all callers' peaks\n",
    "  if (opt$combine_callers == TRUE){\n",
    "    dbObj.total <- dba.peakset(dbObj.noblacklist, consensus=c(DBA_CONDITION, DBA_REPLICATE), minOverlap=1)\n",
    "  }else{\n",
    "    dbObj.total <- dba.peakset(dbObj.noblacklist, consensus=c(DBA_CONDITION, DBA_REPLICATE), minOverlap=0.66)\n",
    "  }\n",
    "  # resulting consensus between callers\n",
    "  dbObj.caller_consensus <- dba(dbObj.total, mask=dbObj.total$masks$Consensus, minOverlap=1)\n",
    "  if (length(unique(dbObj$samples$Replicate)) > 1){\n",
    "    # Get consensus between replicates for each condition (peaks must be overlapping in at least 2 replicates)\n",
    "    dbObj.final <- dba.peakset(dbObj.caller_consensus, consensus=c(DBA_CONDITION), minOverlap=rep_overlaps)\n",
    "    maskname <- names(dbObj.final$masks)[grepl('Replicate.1-2', names(dbObj.final$masks))]\n",
    "    # resulting consensus between replicates\n",
    "    if (length(maskname) == 1){\n",
    "      dbObj.consensus <- dba(dbObj.final, mask=dbObj.final$masks[[maskname]], minOverlap=1)\n",
    "    }else if (length(maskname) == 2){\n",
    "      dbObj.consensus <- dba(dbObj.final, mask=(dbObj.final$masks[[maskname[1]]] | dbObj.final$masks[[maskname[2]]]), minOverlap=1)\n",
    "    }\n",
    "  }else{\n",
    "    dbObj.consensus <- dbObj.caller_consensus\n",
    "  }\n",
    "}else{\n",
    "  if (length(unique(dbObj$samples$Replicate)) > 1){\n",
    "    # consensus between replicates\n",
    "    dbObj.total <- dba.peakset(dbObj.noblacklist, consensus=c(DBA_CONDITION), minOverlap=rep_overlaps)\n",
    "    maskname <- names(dbObj.total$masks)[grepl('Replicate.1-2', names(dbObj.total$masks))]\n",
    "    if (length(maskname) == 1){\n",
    "      dbObj.consensus <- dba(dbObj.total, mask=dbObj.total$masks[[maskname]], minOverlap=1)\n",
    "      dbObj.caller_consensus <- dbObj.consensus\n",
    "    }else if (length(maskname) == 2){\n",
    "      dbObj.consensus <- dba(dbObj.total, mask=(dbObj.total$masks[[maskname[1]]] | dbObj.total$masks[[maskname[2]]]), minOverlap=1)\n",
    "      dbObj.caller_consensus <- dbObj.consensus\n",
    "    }\n",
    "  } else {\n",
    "    # When only 1 peak caller and 1 replicate available\n",
    "    dbObj.total <- dbObj.noblacklist\n",
    "    dbObj.consensus <- dbObj.noblacklist\n",
    "    dbObj.caller_consensus <- dbObj.noblacklist\n",
    "    \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c98c5be",
   "metadata": {},
   "source": [
    "Display caller-consensus peak sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45973481",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dbObj.caller_consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d84fac",
   "metadata": {},
   "source": [
    "Display final consensus peak sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ca458ae",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dbObj.consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71599f42",
   "metadata": {},
   "source": [
    "Re-sort colour code and format attributes for later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd1ac8c8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Add fragment sizes to new objects (helps to later run affinity analysis)\n",
    "for (c in unique(dbObj$samples$Condition)){\n",
    "  for (i in 1:length(dbObj.caller_consensus$mask[[c]])){\n",
    "    if (dbObj.caller_consensus$mask[[c]][[i]]){\n",
    "      dbObj.caller_consensus$config$fragmentSize[i] <- frag_sizes[[c]]\n",
    "    }\n",
    "  }\n",
    "}\n",
    "for (c in unique(dbObj$samples$Condition)){\n",
    "  for (i in 1:length(dbObj.consensus$mask[[c]])){\n",
    "    if (dbObj.consensus$mask[[c]][[i]]){\n",
    "      dbObj.consensus$config$fragmentSize[i] <- frag_sizes[[c]]\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "# Re-sort colours if condition orders changed after consensus (occurs when one condition has only 1 replicate, consensus must be added manually...)\n",
    "# dbObj.consensus <- dba(dbObj.final, mask=(dbObj.final$masks$`Replicate.1-2` | dbObj.final$masks$CONDITION_WITH_ONE_REPLICATE), minOverlap=1)\n",
    "temp <- list()\n",
    "for (i in 1:length(conditions_colour_code)){\n",
    "  temp[names(conditions_colour_code[which(names(conditions_colour_code) == dba.show(dbObj.consensus)$Condition[i])])] <- conditions_colour_code[which(names(conditions_colour_code) == dba.show(dbObj.consensus)$Condition[i])]\n",
    "}\n",
    "conditions_colour_code <- temp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af68b4df",
   "metadata": {},
   "source": [
    "Correlation heatmap of consensus peaksets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969cf67d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "png(paste(supplementary_dir, 'consensus_heatmap.png', sep=''))\n",
    "dba.plotHeatmap(dbObj.consensus)\n",
    "invisible(capture.output( dev.off() ))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6dd89d",
   "metadata": {},
   "source": [
    "Create directories for each condition's outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d2e979",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "result_dirs <- list()\n",
    "for (p in unique(dba.show(dbObj.consensus)$Condition)){\n",
    "  result_dirs[[p]] <- paste(result_dir, p, \"/\", sep='')\n",
    "  if (!file.exists(result_dirs[[p]])) {\n",
    "    dir.create(result_dirs[[p]])\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696d674c",
   "metadata": {},
   "source": [
    "#### Output consensus peaksets to bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3de1a167",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Consensus peaks from all conditions (all relevant peaks) from DiffBind object of final consensus\n",
    "consensus_peaks <- dba.peakset(dbObj.consensus, bRetrieve=TRUE)\n",
    "# Output consensus peaksets\n",
    "raw_peaks <- list()\n",
    "for (c in unique(dba.show(dbObj.consensus)$Condition)){\n",
    "  p <- dba.peakset(dbObj.consensus, dbObj.consensus$masks[[c]], bRetrieve=TRUE)\n",
    "  write.table(as.data.frame(p), file=paste(result_dirs[[c]], c, '_consensus.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "  raw_peaks[[c]] <- p\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83da6ec7",
   "metadata": {},
   "source": [
    "#### Annotate consensus peaksets and output to .tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7abc523",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Get Annotations =========\n",
    "peakAnnoList <- list()\n",
    "for (p in names(raw_peaks)){\n",
    "  cat(\"\\nAnnotating\", p, ' consensus\\n')\n",
    "  anno <- annotatePeak(raw_peaks[[p]], \n",
    "                       TxDb=anno_ref$txdb,\n",
    "                       annoDb=anno_ref$annoDb,\n",
    "                       level=opt$annotation_level,\n",
    "                       tssRegion=c(-3000, 3000)\n",
    "  )\n",
    "  peakAnnoList[[p]] <- anno\n",
    "  write.table(anno@anno, file=paste(result_dirs[[p]], p, '_consensus_annotated.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "778f7f6c",
   "metadata": {},
   "source": [
    "#### Create UpSetPlot of gene sets from annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962284ef",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot UpSet\n",
    "upsetlist <- list()\n",
    "for (p in names(peakAnnoList)) {\n",
    "  upsetlist[[p]] <- peakAnnoList[[p]]@anno$SYMBOL\n",
    "}\n",
    "upset_colors <- list()\n",
    "for (n in names(peakAnnoList)){\n",
    "  upset_colors[[ conditions_colour_code[[n]] ]] <- length(unique(peakAnnoList[[n]]@anno$SYMBOL))\n",
    "}\n",
    "upset_colors <- sort(unlist(upset_colors), decreasing=TRUE)\n",
    "png(paste(result_dir, 'consensus_annotated-genes_upsetplot.png', sep=''),\n",
    "    width = 1920,\n",
    "    height = 1080,\n",
    "    res=300\n",
    ")\n",
    "upset(fromList(upsetlist), \n",
    "      order.by = \"freq\", \n",
    "      nsets = length(names(peakAnnoList)),\n",
    "      sets.bar.color = names(upset_colors),\n",
    "      empty.intersections = \"on\",\n",
    "      set_size.show = TRUE,\n",
    "      set_size.angles = 0,\n",
    "      set_size.scale_max = dim(fromList(upsetlist))[[1]]*1.25,\n",
    "      sets.x.label = \"Gene Set Size\",\n",
    "      mainbar.y.label = \"Intersection Size of Gene Sets\",\n",
    "      mb.ratio = c(0.7, 0.3)\n",
    ") \n",
    "grid.text(\"Consensus Peaksets Annotated Genes\",x = 0.65, y=0.95, gp=gpar(fontsize=10))\n",
    "dev.off()\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "839af6d2",
   "metadata": {},
   "source": [
    "### Obtain differential and shared binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66b2e4f2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Get overlaps\n",
    "differential_peaks <- dba.overlap(dbObj.consensus, dbObj.consensus$masks$All, DataType=DBA_DATA_GRANGES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "918a572d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Differential peaks between conditions (peaks unique to each condition)\n",
    "unique_peaks <- list()\n",
    "for (c in unique(dba.show(dbObj.consensus)$Condition)){\n",
    "  unique_peaks[[c]] <- differential_peaks[[which(unique(dba.show(dbObj.consensus)$Condition) == c)]]\n",
    "  cat(\"\\n\", c, \"has\", length(unique_peaks[[c]]), \"unique peaks.\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cda0c1e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Peaks shared between all conditions\n",
    "shared_peaks <- list()\n",
    "shared_peaks[[\"Shared\"]] <- differential_peaks$inAll\n",
    "cat(\"\\n\", length(shared_peaks[[\"Shared\"]]), \"all shared peaks.\\n\")\n",
    "#write.table(as.data.frame(p), file=paste(result_dirs[[c]], '../Shared_consensus.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "conditions_colour_code[['Shared']] <- \"grey\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "807f378a",
   "metadata": {},
   "source": [
    "Display overlap amounts for each combination (up to 4 conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9678188e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "if (length(unique(dbObj$samples$Condition)) == 3){\n",
    "  for (c in unique(dba.show(dbObj.consensus)$Condition)){\n",
    "    i <- which(unique(dba.show(dbObj.consensus)$Condition) == c) + 3\n",
    "    pair <- unique(dba.show(dbObj.consensus)$Condition)[which(unique(dba.show(dbObj.consensus)$Condition) != c)]\n",
    "    unique_peaks[[paste(pair, collapse='_and_')]] <- differential_peaks[[which(unique(dba.show(dbObj.consensus)$Condition) == c) + 3]]\n",
    "    conditions_colour_code[[paste(pair, collapse='_and_')]] <- colours[i]\n",
    "    cat(\"\\n\", paste(pair, collapse='_and_'), \"have\", length(unique_peaks[[paste(pair, collapse='_and_')]]), \"shared peaks.\\n\")\n",
    "  }\n",
    "}\n",
    "\n",
    "if (length(unique(dbObj$samples$Condition)) == 4){\n",
    "  e = c(\n",
    "    \"A\"=length(unique_peaks[[names(unique_peaks)[1]]]), \n",
    "    \"B\"=length(unique_peaks[[names(unique_peaks)[2]]]),\n",
    "    \"C\"=length(unique_peaks[[names(unique_peaks)[3]]]),\n",
    "    \"D\"=length(unique_peaks[[names(unique_peaks)[4]]]),\n",
    "    \"A&B\"=length(differential_peaks$AandB),\n",
    "    \"A&C\"=length(differential_peaks$AandC),\n",
    "    \"A&D\"=length(differential_peaks$AandD),\n",
    "    \"B&C\"=length(differential_peaks$BandC),\n",
    "    \"B&D\"=length(differential_peaks$BandD),\n",
    "    \"C&D\"=length(differential_peaks$CandD),\n",
    "    \"A&B&C\"=length(differential_peaks$notD),\n",
    "    \"A&B&D\"=length(differential_peaks$notC),\n",
    "    \"A&C&D\"=length(differential_peaks$notB),\n",
    "    \"B&C&D\"=length(differential_peaks$notA),\n",
    "    \"A&B&C&D\"=length(shared_peaks[[names(shared_peaks)[1]]])\n",
    "  )\n",
    "  names(e) = c(names(unique_peaks)[1], names(unique_peaks)[2], names(unique_peaks)[3], names(unique_peaks)[4],\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[3] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep='')\n",
    "  )\n",
    "  for (c in (length(unique(dbObj$samples$Condition))+1):(length(names(e))-1) ){\n",
    "    combo <- str_replace_all(names(e)[c], \"&\", \"_and_\")\n",
    "    unique_peaks[[combo]] <- differential_peaks[[c, ]]\n",
    "    conditions_colour_code[[combo]] <- colours[i]\n",
    "    cat(\"\\n\", combo, \"have\", length(unique_peaks[[combo]]), \"shared peaks.\\n\")\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1c1b3ea",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Put all peaksets into a list\n",
    "peaks <- c(unique_peaks, shared_peaks)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3340ae67",
   "metadata": {},
   "source": [
    "Create directories for outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b010dc19",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for (p in names(peaks)){\n",
    "  result_dirs[[p]] <- paste(result_dir, p, \"/\", sep='')\n",
    "  if (!file.exists(result_dirs[[p]])) {\n",
    "    dir.create(result_dirs[[p]])\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0977f7e0",
   "metadata": {},
   "source": [
    "Output peaksets to bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2271bdb5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for (p in names(peaks)){\n",
    "  write.table(as.data.frame(peaks[[p]]), file=paste(result_dirs[[p]], p, '.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c1cc0f3",
   "metadata": {},
   "source": [
    "#### Plot venn diagram of peaksets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ef66dc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Venn diagram plots\n",
    "if (length(unique(dba.show(dbObj.consensus)$Condition)) == 2){\n",
    "  grid.newpage()\n",
    "  g = draw.pairwise.venn(area1=length(unique_peaks[[names(unique_peaks)[1]]])+length(shared_peaks[[names(shared_peaks)[1]]]), \n",
    "                         area2=length(unique_peaks[[names(unique_peaks)[2]]])+length(shared_peaks[[names(shared_peaks)[1]]]),\n",
    "                         cross.area=length(shared_peaks[[names(shared_peaks)[1]]]),\n",
    "                         category=names(unique_peaks),\n",
    "                         fill=unname(unlist(conditions_colour_code))[1:2],\n",
    "                         col=NA,\n",
    "                         #cat.pos=c(0,0),\n",
    "                         cat.dist = c(0,0))\n",
    "  plt <- grid.arrange(gTree(children=g), top=\"Binding Site Overlaps\", bottom=gsub('/', '', opt$result_dir))\n",
    "  invisible(capture.output(ggsave(filename=paste(result_dir, 'consensus_peaks_venn.png', sep=''), plot=plt)))\n",
    "  rm(g)\n",
    "}else if (length(unique(dba.show(dbObj.consensus)$Condition)) == 3){\n",
    "  e = c(\n",
    "    \"A\"=length(unique_peaks[[names(unique_peaks)[1]]]), \n",
    "    \"B\"=length(unique_peaks[[names(unique_peaks)[2]]]),\n",
    "    \"C\"=length(unique_peaks[[names(unique_peaks)[3]]]),\n",
    "    \"A&B\"=length(differential_peaks$notC),\n",
    "    \"B&C\"=length(differential_peaks$notA),\n",
    "    \"A&C\"=length(differential_peaks$notB),\n",
    "    \"A&B&C\"=length(shared_peaks[[names(shared_peaks)[1]]])\n",
    "  )\n",
    "  names(e) = c(names(unique_peaks)[1], names(unique_peaks)[2], names(unique_peaks)[3],\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&' , names(unique_peaks)[3], sep='')\n",
    "  )\n",
    "  \n",
    "  #png(paste(output_prefix, 'raw_consensus_peaks.png', sep=\"\"))\n",
    "  plt <- plot(euler(e), main=gsub('/', '', opt$result_dir), quantities=TRUE, fills=unname(unlist(conditions_colour_code)))\n",
    "  invisible(capture.output(ggsave(filename=paste(result_dir, 'consensus_peaks_venn.png', sep=''), plot=plt)))\n",
    "  #invisible(capture.output(dev.off()))\n",
    "  rm(e)\n",
    "}else if (length(unique(dba.show(dbObj.consensus)$Condition)) == 4){\n",
    "  e = c(\n",
    "    \"A\"=length(unique_peaks[[names(unique_peaks)[1]]]), \n",
    "    \"B\"=length(unique_peaks[[names(unique_peaks)[2]]]),\n",
    "    \"C\"=length(unique_peaks[[names(unique_peaks)[3]]]),\n",
    "    \"D\"=length(unique_peaks[[names(unique_peaks)[4]]]),\n",
    "    \"A&B\"=length(differential_peaks$AandB),\n",
    "    \"A&C\"=length(differential_peaks$AandC),\n",
    "    \"A&D\"=length(differential_peaks$AandD),\n",
    "    \"B&C\"=length(differential_peaks$BandC),\n",
    "    \"B&D\"=length(differential_peaks$BandD),\n",
    "    \"C&D\"=length(differential_peaks$CandD),\n",
    "    \"A&B&C\"=length(differential_peaks$notD),\n",
    "    \"A&B&D\"=length(differential_peaks$notC),\n",
    "    \"A&C&D\"=length(differential_peaks$notB),\n",
    "    \"B&C&D\"=length(differential_peaks$notA),\n",
    "    \"A&B&C&D\"=length(shared_peaks[[names(shared_peaks)[1]]])\n",
    "  )\n",
    "  names(e) = c(names(unique_peaks)[1], names(unique_peaks)[2], names(unique_peaks)[3], names(unique_peaks)[4],\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[3] , '&' , names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[3], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[2] , '&' , names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep=''),\n",
    "               paste(names(unique_peaks)[1] , '&' , names(unique_peaks)[2] , '&', names(unique_peaks)[3] , '&', names(unique_peaks)[4], sep='')\n",
    "  )\n",
    "  \n",
    "  #png(paste(output_prefix, 'raw_consensus_peaks.png', sep=\"\"))\n",
    "  plt <- plot(euler(e), main=gsub('/', '', opt$result_dir), quantities=TRUE, fills=unname(unlist(conditions_colour_code)))\n",
    "  invisible(capture.output(ggsave(filename=paste(result_dir, 'consensus_peaks_venn.png', sep=''), plot=plt)))\n",
    "  #invisible(capture.output(dev.off()))\n",
    "  rm(e)\n",
    "}\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b953d8fd",
   "metadata": {},
   "source": [
    "#### Plot PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ca249d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# PCA plot\n",
    "tryCatch(\n",
    "  {\n",
    "    plt <- dba.plotPCA(dbObj, masks=!dbObj.total$masks$Consensus, attributes=DBA_CONDITION, label=DBA_ID, vColors=(colours))\n",
    "    #plt$main <- \"PCA\"\n",
    "    invisible(capture.output(ggsave(filename=paste(supplementary_dir, 'pca_condition.png', sep=''), plot=grid.arrange(plt))))\n",
    "    \n",
    "    if(length(unique(dbObj$samples$Factor)) > 1){\n",
    "      plt <- dba.plotPCA(dbObj, masks=!dbObj.total$masks$Consensus, attributes=DBA_FACTOR, label=DBA_ID)\n",
    "      invisible(capture.output( ggsave(filename=paste(supplementary_dir, 'pca_factor.png', sep=''), plot=grid.arrange(plt)) ))\n",
    "      #invisible(capture.output(dev.off()))\n",
    "    }\n",
    "  },error = function(e)\n",
    "  {\n",
    "    message(e)\n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f12951a1",
   "metadata": {},
   "source": [
    "#### Plot peaks over genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c48ce87",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot peaks over genome\n",
    "tryCatch(\n",
    "  {\n",
    "    plt <- covplot(c(unique_peaks[1:length(unique(dbObj$samples$Condition))], shared_peaks), title=\"Peaks over Genome\") + \n",
    "      scale_color_manual(values=rev(c(unlist(unname(conditions_colour_code[1:length(unique(dbObj$samples$Condition))])), 'grey'))) + \n",
    "      scale_fill_manual(values=rev(c(unlist(unname(conditions_colour_code[1:length(unique(dbObj$samples$Condition))])), 'grey')))\n",
    "    invisible(capture.output(ggsave(filename=paste(result_dir, 'genome_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "    plt <- plt + facet_grid(chr ~ .id)\n",
    "    invisible(capture.output(ggsave(filename=paste(result_dir, 'genome_peaks_split.png', sep=''), plot=plt, dpi=320)))\n",
    "    rm(plt)\n",
    "  },error = function(e)\n",
    "  {\n",
    "    message(e)\n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfed7a44",
   "metadata": {},
   "source": [
    "#### Plot heatmaps of peaksets at TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e43c31b9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot peaks related to TSSites\n",
    "tryCatch(\n",
    "  {\n",
    "    tagMatrices <- list()\n",
    "    for (p in names(peaks)){\n",
    "      tagMatrix <- getTagMatrix(peaks[[p]], windows=promoters)\n",
    "      if (length(tagMatrix) == 0){\n",
    "        cat(\"No peaks at promoter sites for\", p, \"\\n\")\n",
    "        rm(tagMatrix)\n",
    "        #break\n",
    "      }else{\n",
    "        tagMatrices[[p]] <- tagMatrix\n",
    "        cat(dim(tagMatrix)[[1]], \"peaks at promoter sites for\", p, \"\\n\")\n",
    "        plt <- tagHeatmap(tagMatrix, \n",
    "                          xlab=\"bp at TSS\", \n",
    "                          ylab=\"Peaks\", \n",
    "                          title=paste(dim(tagMatrix)[[1]],'Peaks at Promoters', p, sep=\" - \"),\n",
    "                          palette=if_else(conditions_colour_code[[p]] == \"#00BFC4\", 'Greens', 'Reds'), \n",
    "        )\n",
    "        invisible(capture.output(ggsave(paste(result_dirs[[p]], 'TSS_heatmap_', p, '_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "        rm(tagMatrix)\n",
    "        invisible(capture.output(gc()))\n",
    "      }\n",
    "    }\n",
    "  },error = function(e)\n",
    "  {\n",
    "    message(e)\n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad98960",
   "metadata": {},
   "source": [
    "#### Plot average profile of peaksets at TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe28326",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot TSS profile of peaks\n",
    "tryCatch(\n",
    "  {\n",
    "    if (length(unique(dba.show(dbObj.consensus)$Condition)) == 3){\n",
    "      plt <- plotAvgProf(tagMatrices[1:3], xlim=c(-3000, 3000), conf=0.95, resample=1000, ncpus = parallel::detectCores()/2) +\n",
    "        scale_color_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[1:3]) +\n",
    "        scale_fill_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[1:3])\n",
    "      invisible(capture.output(ggsave(paste(result_dir, 'TSS_profile_unique-peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "      invisible(capture.output(gc()))\n",
    "      plt <- plotAvgProf(tagMatrices[4:6], xlim=c(-3000, 3000), conf=0.95, resample=1000, ncpus = parallel::detectCores()/2) +\n",
    "        scale_color_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[4:6]) +\n",
    "        scale_fill_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[4:6])\n",
    "      invisible(capture.output(ggsave(paste(result_dir, 'TSS_profile_pairs-peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "      invisible(capture.output(gc()))\n",
    "    }\n",
    "    else{\n",
    "      plt <- plotAvgProf(tagMatrices, xlim=c(-3000, 3000), conf=0.95, resample=1000, ncpus = parallel::detectCores()/2) +\n",
    "        scale_color_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))) +\n",
    "        scale_fill_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)])))\n",
    "      invisible(capture.output(ggsave(paste(result_dir, 'TSS_profile_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "      invisible(capture.output(gc()))\n",
    "    }\n",
    "    plt <- plotAvgProf(tagMatrices[['Shared']], xlim=c(-3000, 3000), conf=0.95, resample=1000, ncpus = parallel::detectCores()/2) +\n",
    "      scale_color_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[length(conditions_colour_code)]) +\n",
    "      scale_fill_manual(values=unname(unlist(conditions_colour_code[names(tagMatrices)]))[length(conditions_colour_code)])\n",
    "    invisible(capture.output(ggsave(paste(result_dir, 'TSS_profile_shared-peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "    invisible(capture.output(gc()))\n",
    "    rm(tagMatrices)\n",
    "    rm(plt)\n",
    "  },error = function(e)\n",
    "  {\n",
    "    message(e)\n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13fbf77",
   "metadata": {},
   "source": [
    "### Annotate peaksets and perform functional pathway enrichment analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9f7733e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Get Annotations =========\n",
    "peakAnnoList <- list()\n",
    "for (p in names(peaks)){\n",
    "  cat(\"\\nObtaining annotations for \", p, \"\\n\")\n",
    "  \n",
    "  if (p == \"Shared\"){\n",
    "    colour <- \"#56B1F7\"\n",
    "    heat_colour <- \"BuPu\"\n",
    "  }else{\n",
    "    colour <- conditions_colour_code[[p]]\n",
    "    heat_colour <- conditions_colour_code[[p]]\n",
    "  }\n",
    "  \n",
    "  cat(\"\\nAnnotating\", p, '\\n')\n",
    "  anno <- annotatePeak(peaks[[p]], \n",
    "                       TxDb=anno_ref$txdb,\n",
    "                       annoDb=anno_ref$annoDb,\n",
    "                       level=opt$annotation_level,\n",
    "                       tssRegion=c(-3000, 3000)\n",
    "  )\n",
    "  \n",
    "  peakAnnoList[[p]] <- anno\n",
    "  \n",
    "  # Mapper for EntrezID to gene SYMBOL\n",
    "  df <- as.data.frame(anno@anno)\n",
    "  mapper <- df[c('geneId', 'SYMBOL')]\n",
    "  mapper <- mapper[!duplicated(mapper), ]\n",
    "  \n",
    "  plt <- upsetplot(anno, vennpie=TRUE) + ggtitle(p)\n",
    "  invisible(capture.output(ggsave(paste(result_dirs[[p]], p, '_annotated_peaks_upsetplot.png', sep=''), plot=plt, dpi=320, bg='white')))\n",
    "  \n",
    "  tryCatch(\n",
    "    {\n",
    "      png(paste(result_dirs[[p]], p, '_peaks_annotation_pie.png', sep=''), width=1680, height=1200)\n",
    "      plt <- plotAnnoPie(anno, main=paste(p, '\\n\\n', length(anno@anno), ' Sites', sep=''), line=-10, cex.main=3.25, cex=3)\n",
    "      invisible(capture.output( dev.off() ))\n",
    "      #invisible(capture.output( ggsave(filename=paste(result_dirs[[p]], p, '_peaks_annotation_pie.png', sep=''), plot=grid.arrange(plt)) ))\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "    }\n",
    "  )\n",
    "  \n",
    "  # Write annotation to file\n",
    "  write.table(anno, file=paste(result_dirs[[p]], p, '_annotated.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "  \n",
    "  tryCatch(\n",
    "    {\n",
    "      genes <- list()\n",
    "      genes[[p]] <- anno@anno$geneId\n",
    "      names(genes) = sub(\"_\", \"\\n\", names(genes))\n",
    "      cat(\"\\nGetting\", p, 'KEGG\\n')\n",
    "      \n",
    "      # fun is \"groupGO\", \"enrichGO\", \"enrichKEGG\", \"enrichDO\" or \"enrichPathway\" \n",
    "      compKEGG <- compareCluster(geneCluster=genes,\n",
    "                                 fun=\"enrichKEGG\",\n",
    "                                 pvalueCutoff=opt$fdr,\n",
    "                                 pAdjustMethod=\"BH\",\n",
    "                                 minGSSize = opt$minGSSize,\n",
    "                                 maxGSSize = opt$maxGSSize,\n",
    "                                 organism=anno_ref$keggOrg\n",
    "      ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "      if (!class(compKEGG) == 'compareClusterResult'){\n",
    "        cat(\"\\nNo results.\\n\")\n",
    "        #next\n",
    "      }else{\n",
    "        cat('\\n', dim(compKEGG@compareClusterResult)[1], 'results\\n')\n",
    "      }\n",
    "      if ((!is.null(compKEGG)) & (dim(compKEGG@compareClusterResult)[1] > 0)){\n",
    "        # Map EntrezIDs to gene SYMBOL\n",
    "        compKEGG@compareClusterResult$SYMBOL <- compKEGG@compareClusterResult$geneID\n",
    "        myEntrez <- lapply(compKEGG@compareClusterResult$geneID, strsplit, '/')\n",
    "        for (i in 1:length(myEntrez)){\n",
    "          compKEGG@compareClusterResult$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "        }\n",
    "        # Write annotations to csv\n",
    "        write.table(as.data.frame(compKEGG), file=paste(result_dirs[[p]], p, '_annotated_KEGG.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "        \n",
    "        plt <- make_dotplot(compKEGG@compareClusterResult, title=paste('KEGG - ', p, sep=\"\"), ylabel=\"KEGG Category\", colour=colour, n=15)\n",
    "        invisible(capture.output(ggsave(filename=paste(result_dirs[[p]], p, '_annotated_KEGG.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "        remove(plt)\n",
    "        remove(compKEGG)\n",
    "        gc()\n",
    "      } else{\n",
    "        cat(\"\\nNo annotation results\\n\")\n",
    "        remove(compKEGG)\n",
    "        gc()\n",
    "      }\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "      remove(compKEGG)\n",
    "      gc()\n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "  for (ont in c('ALL', 'CC', 'MF', 'BP')){\n",
    "    tryCatch(\n",
    "      {\n",
    "        genes <- list()\n",
    "        genes[[p]] <- anno@anno$SYMBOL\n",
    "        names(genes) = sub(\"_\", \"\\n\", names(genes))\n",
    "        cat(\"\\nGetting \", p, ' GO ', ont, '\\n')\n",
    "        \n",
    "        compGO <- compareCluster(geneCluster=genes,\n",
    "                                 keyType='SYMBOL',\n",
    "                                 OrgDb=anno_ref$annoDb,\n",
    "                                 fun=\"enrichGO\",\n",
    "                                 ont=ont,\n",
    "                                 pvalueCutoff=opt$fdr,\n",
    "                                 pAdjustMethod=\"BH\",\n",
    "                                 minGSSize = opt$minGSSize,\n",
    "                                 maxGSSize = opt$maxGSSize,\n",
    "                                 readable=TRUE\n",
    "        ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "        if (!class(compGO) == 'compareClusterResult'){\n",
    "          cat(\"\\nNo results.\\n\")\n",
    "          next\n",
    "        }else{\n",
    "          cat('\\n', dim(compGO@compareClusterResult)[1], 'results\\n')\n",
    "        }\n",
    "        if ((!is.null(compGO)) & (dim(compGO@compareClusterResult)[1] > 0)){\n",
    "          compGO@compareClusterResult$ONTOLOGY <- go2ont(compGO@compareClusterResult$ID)$Ontology\n",
    "          # Write annotations to csv\n",
    "          write.table(as.data.frame(compGO), file=paste(result_dirs[[p]], p, '_annotated_GO-', ont, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "          \n",
    "          plt <- make_dotplot(compGO@compareClusterResult, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\", colour=colour, n=15)\n",
    "          invisible(capture.output(ggsave(filename=paste(result_dirs[[p]], p, '_annotated_GO-', ont, '.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "          remove(plt)\n",
    "          remove(compGO)\n",
    "          gc()\n",
    "        } else{\n",
    "          cat(\"\\nNo annotation results\\n\")\n",
    "          remove(compGO)\n",
    "          gc()\n",
    "        }\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "        remove(compGO)\n",
    "        gc()\n",
    "      }\n",
    "    )\n",
    "    invisible(capture.output(gc()))\n",
    "  }\n",
    "  \n",
    "  genes_entrez <- list()\n",
    "  genes_entrez[[p]] <- anno@anno$geneId\n",
    "  names(genes_entrez) = sub(\"_\", \"\\n\", names(genes_entrez))\n",
    "  for (annotation_type in c(\"GOTERM_BP_DIRECT\", \"GOTERM_CC_DIRECT\", \"GOTERM_MF_DIRECT\", \"KEGG_PATHWAY\")){\n",
    "    cat(\"\\nGetting \", p, ' DAVID ', annotation_type, ' annotations...\\n')  \n",
    "    # DAVID Annotation\n",
    "    tryCatch(\n",
    "      {\n",
    "        cat(\"\\nDAVID - \", annotation_type, \"\\n\")\n",
    "        entrez <- unname(genes_entrez[[p]][!is.na(unname(genes_entrez[[p]]))])\n",
    "        \n",
    "        py_run_string(\"import pandas as pd\")\n",
    "        py_run_string(\"import sys\")\n",
    "        py_run_string(\"from suds.client import Client\")\n",
    "        \n",
    "        # create a service client using the wsdl.\n",
    "        py_run_string(\"client = Client('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService?wsdl')\")\n",
    "        py_run_string(\"client.wsdl.services[0].setlocation('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService.DAVIDWebServiceHttpSoap11Endpoint/')\")\n",
    "        \n",
    "        #authenticate user email\n",
    "        py_run_string(\"client.service.authenticate(r.opt['david_user'])\")\n",
    "        \n",
    "        # Read input gene list file, convert ids to a comma-delimited string and upload the list to DAVID\n",
    "        py_run_string(\"inputIds = ','.join(r.entrez)\")\n",
    "        py_run_string(\"client.service.addList(inputIds, 'ENTREZ_GENE_ID', r.p, 0)\")\n",
    "        \n",
    "        # setCategories\n",
    "        py_run_string(\"categorySting = str(client.service.setCategories(r.annotation_type))\")\n",
    "        \n",
    "        #getChartReport\n",
    "        py_run_string(\"thd = r.opt['fdr']\")\n",
    "        py_run_string(\"ct = r.opt['minGSSize']\")\n",
    "        py_run_string(\"chartReport = client.service.getChartReport(thd,ct)\")\n",
    "        py_run_string(\"chartRow = len(chartReport)\")\n",
    "        py_run_string(\"print ('Total chart records:',chartRow)\")\n",
    "        \n",
    "        if (py$chartRow > 0){\n",
    "          if (annotation_type == \"KEGG_PATHWAY\"){\n",
    "            splitter <- \":\"\n",
    "          }else{\n",
    "            splitter <- \"~\"\n",
    "          }\n",
    "          # parse chartReport\n",
    "          py_run_string(\"records = pd.DataFrame()\")\n",
    "          py_run_string(\"for simpleChartRecord in chartReport:\n",
    "                          df = pd.DataFrame(index=[records.shape[0]], data={\n",
    "                              'ID' : simpleChartRecord['termName'].split(r.splitter)[0],\n",
    "                              'Category' : simpleChartRecord['categoryName'],\n",
    "                              'Description' : simpleChartRecord['termName'].split(r.splitter)[1],\n",
    "                              'GeneRatio': str(simpleChartRecord['listHits']) + '/' + str(simpleChartRecord['listTotals']), \n",
    "                              'BgRatio': str(simpleChartRecord['popHits']) + '/' + str(simpleChartRecord['popTotals']), \n",
    "                              'pvalue' : simpleChartRecord['ease'],\n",
    "                              'p.adjust' : simpleChartRecord['benjamini'],\n",
    "                              'FDR' : simpleChartRecord['afdr'],\n",
    "                              'geneID' : simpleChartRecord['geneIds'].replace(', ', '/'),\n",
    "                              'Count' : simpleChartRecord['listHits'],\n",
    "                              'foldEnrichment' : simpleChartRecord['foldEnrichment'],\n",
    "                              'id' : simpleChartRecord['id']\n",
    "                              })\n",
    "                          records = pd.concat([records, df])\")\n",
    "          #py_run_string(\"records.reset_index(drop=True, inplace=True)\")\n",
    "          compD <- py$records\n",
    "          \n",
    "          qobj <- tryCatch(qvalue(p=as.numeric(compD$pvalue), lambda=opt$pvalue, pi0.method=\"bootstrap\"),\n",
    "                           error=function(e) NULL)\n",
    "          if (class(qobj) == \"qvalue\") {\n",
    "            qvalues <- qobj$qvalues\n",
    "          } else {\n",
    "            qvalues <- NA\n",
    "          }\n",
    "          compD$qvalue <- qvalues\n",
    "          \n",
    "          compDAVID <- new(\"enrichResult\",\n",
    "                           result         = compD,\n",
    "                           pvalueCutoff   = 1,\n",
    "                           pAdjustMethod  = \"BH\",\n",
    "                           organism       = opt$assembly,\n",
    "                           ontology       = annotation_type,\n",
    "                           gene           = entrez,\n",
    "                           keytype        = \"ENTREZ_GENE_ID\")\n",
    "          rm(compD)\n",
    "          \n",
    "          if (!class(compDAVID) == 'enrichResult'){\n",
    "            cat(\"\\nNo results.\\n\")\n",
    "            next\n",
    "          }else{\n",
    "            cat('\\n', dim(compDAVID@result)[1], 'results\\n')\n",
    "          }\n",
    "          \n",
    "          if ((!is.null(compDAVID)) & (dim(compDAVID@result)[1] > 0)){\n",
    "            # Map EntrezIDs to gene SYMBOL\n",
    "            compDAVID@result$SYMBOL <- compDAVID@result$geneID\n",
    "            myEntrez <- lapply(compDAVID@result$geneID, strsplit, '/')\n",
    "            for (i in 1:length(myEntrez)){\n",
    "              compDAVID@result$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "            }\n",
    "            # Write annotations to csv\n",
    "            write.table(as.data.frame(compDAVID), file=paste(result_dirs[[p]], 'DAVID_annotation_', annotation_type, '_', p, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "            \n",
    "            plt <- make_dotplot(compDAVID@result, title=paste('DAVID - ', p, sep=\"\"), ylabel=paste(annotation_type,\"Category\", sep=' '), colour=colour, n=15)\n",
    "            invisible(capture.output(ggsave(filename=paste(result_dirs[[p]], 'DAVID_annotation_', annotation_type, '_', p, '_dotplot.png', sep=''), plot=plt, dpi=320)))\n",
    "            remove(plt)\n",
    "            remove(compDAVID)\n",
    "            gc()\n",
    "          } else{\n",
    "            cat(\"\\nNo annotation results\\n\")\n",
    "            remove(compDAVID)\n",
    "            gc()\n",
    "          }\n",
    "        }\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "        remove(compDAVID)\n",
    "        gc()\n",
    "      }\n",
    "    )\n",
    "  }\n",
    "    \n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- plotAnnoBar(peakAnnoList)\n",
    "      invisible(capture.output(ggsave(filename=paste(result_dir, 'peaks_annotation_distribution_bar.png', sep=''), plot=plt, dpi=320)))\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "    }\n",
    "  )\n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- plotDistToTSS(peakAnnoList)\n",
    "      invisible(capture.output(ggsave(filename=paste(result_dir, 'peaks_annotation_TSS_distribution.png', sep=''), plot=plt, dpi=320)))\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cce4d1c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Free up some memory\n",
    "rm(dbObj.final)\n",
    "rm(dbObj.total)\n",
    "rm(plt)\n",
    "rm(anno)\n",
    "rm(peakAnnoList)\n",
    "rm(compKEGG)\n",
    "rm(compGO)\n",
    "invisible(capture.output(gc()))\n",
    "\n",
    "# Exit script if done\n",
    "if (opt$occupancy_only){\n",
    "  cat(\"Skipping affinity analysis.\\nDone!\\n\")\n",
    "  q()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0230c7c5",
   "metadata": {},
   "source": [
    "## <center> =========== END OF OCCUPANCY ANALYSIS ============ </center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b103ae5c",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83e6e990",
   "metadata": {},
   "source": [
    "## <center> =========== Start Affinity Analysis ============ </center>\n",
    "The consensus peaks determined above are used here to focus on relevant peak regions only. This way, read\n",
    "counts for consensus peaks are analyzed for statistical significance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "527bfcd2",
   "metadata": {},
   "source": [
    "Set directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e47f73",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat(\"\\n=====  Affinity Analysis =====\\n\")\n",
    "result_dir <- paste(opt$result_dir, 'Affinity_Analysis/', sep='')\n",
    "# Make new directory for analysis\n",
    "if (!file.exists(result_dir)) {\n",
    "  dir.create(result_dir)\n",
    "}\n",
    "\n",
    "supplementary_dir <- paste(result_dir, \"Supplementary/\", sep='')\n",
    "if (!file.exists(supplementary_dir)) {\n",
    "  dir.create(supplementary_dir)\n",
    "}\n",
    "\n",
    "if (!file.exists(paste(result_dir, 'DESeq2/', sep=''))) {\n",
    "  dir.create(paste(result_dir, 'DESeq2/', sep=''))\n",
    "}\n",
    "if (!file.exists(paste(result_dir, 'edgeR/', sep=''))) {\n",
    "  dir.create(paste(result_dir, 'edgeR/', sep=''))\n",
    "}\n",
    "\n",
    "output_prefix <- change_dirs(result_dir, '', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c13f15b7",
   "metadata": {},
   "source": [
    "### Count reads for consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ca0d01",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Count Reads from bams for All Relevant Consensus Peaks =========\n",
    "#(recommend to use ‘summits=100‘ for ATAC-seq).\n",
    "\n",
    "# If only 1 peak caller in samplesheet, use the caller_consensus DiffBind object\n",
    "if (length(unique(dbObj$samples$Factor)) > 1){\n",
    "  for (n in names(dbObj.caller_consensus$config)){\n",
    "    if (n != 'fragmentSize'){\n",
    "      dbObj.caller_consensus$config[[n]] <- unique(dbObj.caller_consensus$config[[n]])\n",
    "    }\n",
    "  }\n",
    "  dbObj.counted <- dba.count(dbObj.caller_consensus, bUseSummarizeOverlaps=TRUE,\n",
    "                             peaks=consensus_peaks,\n",
    "                             minOverlap=1,\n",
    "                             score=DBA_SCORE_NORMALIZED,\n",
    "                             bParallel=TRUE,\n",
    "                             #fragmentSize=dbObj.caller_consensus$config$fragmentSize,\n",
    "                             #summits=200, filter=1, bRemoveDuplicates=FALSE, bScaleControl=TRUE,\n",
    "                             #bSubControl=is.null(dbObj.noblacklist$greylist),\n",
    "                             #mapQCth=dbObj.noblacklist$config$mapQCth,\n",
    "                             #filterFun=max, minCount=0,\n",
    "                             #bLog=FALSE,\n",
    "                             #readFormat=DBA_READS_DEFAULT,\n",
    "                             #bParallel=dbObj.noblacklist$config$RunParallel\n",
    "  )\n",
    "} else { # Otherwise, use the raw peaksets with blacklisted regions removed\n",
    "  for (n in names(dbObj.noblacklist$config)){\n",
    "    if (n != 'fragmentSize'){\n",
    "      dbObj.noblacklist$config[[n]] <- unique(dbObj.noblacklist$config[[n]])\n",
    "    }\n",
    "  }\n",
    "  dbObj.counted <- dba.count(dbObj.noblacklist, bUseSummarizeOverlaps=TRUE,\n",
    "                             peaks=consensus_peaks, \n",
    "                             minOverlap=1, \n",
    "                             score=DBA_SCORE_NORMALIZED,\n",
    "                             bParallel=TRUE,\n",
    "                             #fragmentSize=dbObj.noblacklist$config$fragmentSize,\n",
    "                             #summits=200, filter=1, bRemoveDuplicates=FALSE, bScaleControl=TRUE,\n",
    "                             #bSubControl=is.null(dbObj.noblacklist$greylist),\n",
    "                             #mapQCth=dbObj.noblacklist$config$mapQCth,\n",
    "                             #filterFun=max, minCount=0,\n",
    "                             #bLog=FALSE,\n",
    "                             #readFormat=DBA_READS_DEFAULT,\n",
    "                             #bParallel=dbObj.noblacklist$config$RunParallel\n",
    "  )\n",
    "}\n",
    "dbObj.counted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d89378",
   "metadata": {},
   "source": [
    "### Normalize counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e88eb9f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Normalize Counts =========\n",
    "tryCatch(\n",
    "  {\n",
    "    # Normalize (\"safest general method\")\n",
    "    dbObj.norm <- dba.normalize(dbObj.counted, method=DBA_ALL_METHODS, \n",
    "                                normalize=DBA_NORM_NATIVE,\n",
    "                                background=TRUE, library=DBA_LIBSIZE_DEFAULT,\n",
    "                                spikein=FALSE, offsets=FALSE,\n",
    "                                libFun=mean, bRetrieve=FALSE)\n",
    "  },error = function(e)\n",
    "  { \n",
    "    message(e, \"\\nUsing approximate normalization...\")\n",
    "    # Approximate normalization to above without extra reading of bam files\n",
    "    dbObj.norm <- dba.normalize(dbObj.counted, method=DBA_ALL_METHODS, \n",
    "                                normalize=DBA_NORM_LIB,\n",
    "                                background=FALSE, library=DBA_LIBSIZE_FULL,\n",
    "                                spikein=FALSE, offsets=FALSE,\n",
    "                                libFun=mean, bRetrieve=FALSE)\n",
    "  }\n",
    ")\n",
    "cat(\"After normalizing:\\n\")\n",
    "dbObj.norm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83c94e71",
   "metadata": {},
   "source": [
    "### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fafc695",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Correlation heatmap\n",
    "png(paste(supplementary_dir, 'consensus_peaks_counted_normalized_heatmap.png', sep=''))\n",
    "dba.plotHeatmap(dbObj.norm)\n",
    "invisible(capture.output( dev.off() ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e341e5df",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# PCA plot\n",
    "png(paste(supplementary_dir, 'consensus_peaks_counted_normalized_pca.png', sep=''))\n",
    "dba.plotPCA(dbObj.norm, attributes=DBA_CONDITION, label=DBA_ID, vColors=(colours))\n",
    "invisible(capture.output( dev.off() ))\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae2b77c1",
   "metadata": {},
   "source": [
    "### Define contrasts for differential comparison\n",
    "May need to manually select 2 conditions if more than 2 in samplesheet/dbObj  \n",
    "See DiffBind docs for more complex contrast comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ca46fd1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Define contrasts between sample conditions (assumes 2 conditions) =========\n",
    "# Define design and contrasts explicitly\n",
    "if (length(unique(dbObj.norm$samples$Replicate)) > 1) {\n",
    "  dbObj.contrast <- dba.contrast(dbObj.norm, design=\"~Condition\", \n",
    "                                 minMembers=2, \n",
    "                                 contrast=c(\"Condition\", unique(dbObj.norm$samples$Condition)))\n",
    "}else{\n",
    "  # Pre-DiffBind version 3 (or if no use of multiple replicates...will not be able to perform analysis in this case)\n",
    "  dbObj.contrast <- dba.contrast(dbObj.norm, design=FALSE, categories=DBA_CONDITION,\n",
    "                                 group1=dbObj.norm$masks[[unique(dbObj.norm$samples$Condition)[1]]], \n",
    "                                 name1=unique(dbObj.norm$samples$Condition)[1],\n",
    "                                 name2=unique(dbObj.norm$samples$Condition)[2],\n",
    "                                 minMembers=2,\n",
    "                                 reorderMeta=list(Condition=unique(dbObj.norm$samples$Condition)[1], \n",
    "                                                  Condition=unique(dbObj.norm$samples$Condition)[2])\n",
    "  )\n",
    "}\n",
    "cat(\"After defining contrast:\\n\")\n",
    "dbObj.contrast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92a1178b",
   "metadata": {},
   "source": [
    "### Analyze binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa23c30",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ========= Analyze Affinities =========\n",
    "dbObj.analyzed <- dba.analyze(dbObj.contrast, method=DBA_ALL_METHODS, bParallel=TRUE)\n",
    "cat(\"After analyzing:\\n\")\n",
    "dbObj.analyzed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "302125f0",
   "metadata": {},
   "source": [
    "### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a0fd50",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Venn diagram\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(result_dir, 'analyzed_venn.png', sep=''))\n",
    "    plt <- dba.plotVenn(main=\"DE Binding Sites Identified by Method\",\n",
    "                        dbObj.analyzed, \n",
    "                        contrast=1, \n",
    "                        method=DBA_ALL_METHODS\n",
    "    )\n",
    "    invisible(capture.output( dev.off() ))\n",
    "    #invisible(capture.output( ggsave(filename=paste(result_dir, 'analyzed_venn.png', sep=''), plot=grid.arrange(plt)) ))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    if (file.exists(paste(result_dir, 'analyzed_venn.png', sep=''))){\n",
    "      invisible(file.remove(paste(result_dir, 'analyzed_venn.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "723a113c",
   "metadata": {},
   "source": [
    "#### Plots using DESeq2 analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a79772",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Set method to DESeq2\n",
    "m <- DBA_DESEQ2\n",
    "output_prefix <- change_dirs(result_dir, m, '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e4b4d82",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Venn diagram\n",
    "tryCatch(\n",
    "  {\n",
    "    tryCatch(\n",
    "      {\n",
    "        v <- dba.plotVenn(dbObj.analyzed, method=m, contrast=1, \n",
    "                          bDB=TRUE, bGain=TRUE, bLoss=TRUE, bAll=FALSE,\n",
    "                          main=paste(\"Binding Sites - \", m, sep=''))\n",
    "        png(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))\n",
    "        grid.newpage()\n",
    "        g = draw.pairwise.venn(area1=length(v$onlyA)+length(v$inAll), \n",
    "                                area2=length(v$onlyB)+length(v$inAll),\n",
    "                                cross.area=length(v$inAll),\n",
    "                                category=names(unique_peaks),\n",
    "                                fill=unname(unlist(conditions_colour_code)),\n",
    "                                col=NA, cex=1.5, cat.cex=1.5,\n",
    "                                cat.dist = c(0,0))\n",
    "        grid.arrange(gTree(children=g), top=textGrob(\"Binding Sites\", gp=gpar(cex=1.5, fontface='bold')),\n",
    "                      bottom=textGrob(m, gp=gpar(cex=1.5)))\n",
    "        \n",
    "        invisible(capture.output(dev.off()))\n",
    "      }, error=function(e){\n",
    "        message('Peaksets do not meet specified criteria for venn')\n",
    "        \n",
    "        tryCatch(\n",
    "          {\n",
    "            png(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))\n",
    "            v <- dba.plotVenn(dbObj.analyzed, method=m, contrast=1, \n",
    "                              bDB=TRUE, bGain=TRUE, bLoss=TRUE, bAll=TRUE,\n",
    "                              main=paste(\"Binding Sites - \", m, sep=''))\n",
    "            invisible(capture.output(dev.off()))\n",
    "          }, error=function(e){\n",
    "            message('Peaksets do not meet specified criteria for venn')\n",
    "          }\n",
    "        )\n",
    "      }\n",
    "    )\n",
    "    \n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_venn_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0b9cee",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Volcano plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep=''))\n",
    "    par(mfrow=c(1,2))\n",
    "    dba.plotVolcano(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bcbc6c9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# MA plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_ma_', m, '.png', sep=''))\n",
    "    dba.plotMA(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_ma_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_ma_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80bd074b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Box plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_box_', m, '.png', sep=''))\n",
    "    dba.plotBox(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_box_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_box_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f99c898",
   "metadata": {},
   "source": [
    "#### Plots using edgeR analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fbb161a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Set method to edgeR\n",
    "m <- DBA_EDGER\n",
    "output_prefix <- change_dirs(result_dir, m, '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf59c1d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Venn diagram\n",
    "tryCatch(\n",
    "  {\n",
    "    tryCatch(\n",
    "      {\n",
    "        v <- dba.plotVenn(dbObj.analyzed, method=m, contrast=1, \n",
    "                          bDB=TRUE, bGain=TRUE, bLoss=TRUE, bAll=FALSE,\n",
    "                          main=paste(\"Binding Sites - \", m, sep=''))\n",
    "        png(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))\n",
    "        grid.newpage()\n",
    "        g = draw.pairwise.venn(area1=length(v$onlyA)+length(v$inAll), \n",
    "                                area2=length(v$onlyB)+length(v$inAll),\n",
    "                                cross.area=length(v$inAll),\n",
    "                                category=names(unique_peaks),\n",
    "                                fill=unname(unlist(conditions_colour_code)),\n",
    "                                col=NA, cex=1.5, cat.cex=1.5,\n",
    "                                cat.dist = c(0,0))\n",
    "        grid.arrange(gTree(children=g), top=textGrob(\"Binding Sites\", gp=gpar(cex=1.5, fontface='bold')),\n",
    "                      bottom=textGrob(m, gp=gpar(cex=1.5)))\n",
    "        \n",
    "        invisible(capture.output(dev.off()))\n",
    "      }, error=function(e){\n",
    "        message('Peaksets do not meet specified criteria for venn')\n",
    "        \n",
    "        tryCatch(\n",
    "          {\n",
    "            png(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))\n",
    "            v <- dba.plotVenn(dbObj.analyzed, method=m, contrast=1, \n",
    "                              bDB=TRUE, bGain=TRUE, bLoss=TRUE, bAll=TRUE,\n",
    "                              main=paste(\"Binding Sites - \", m, sep=''))\n",
    "            invisible(capture.output(dev.off()))\n",
    "          }, error=function(e){\n",
    "            message('Peaksets do not meet specified criteria for venn')\n",
    "          }\n",
    "        )\n",
    "      }\n",
    "    )\n",
    "    \n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_venn_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_venn_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfe4e04",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Volcano plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep=''))\n",
    "    par(mfrow=c(1,2))\n",
    "    dba.plotVolcano(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_volcano_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e7c2d3e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# MA plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_ma_', m, '.png', sep=''))\n",
    "    dba.plotMA(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_ma_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_ma_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35e24bd8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Box plot\n",
    "tryCatch(\n",
    "  {\n",
    "    png(paste(output_prefix, 'analyzed_box_', m, '.png', sep=''))\n",
    "    dba.plotBox(dbObj.analyzed, method=m)\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_box_', m, '.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_box_', m, '.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a68882",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Reset to main output directory\n",
    "output_prefix <- change_dirs(result_dir, '', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be189d94",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot profile\n",
    "tryCatch(\n",
    "  {\n",
    "    profile_colors <- list()\n",
    "    for (i in 1:length(unique(dbObj.analyzed$samples$Condition))) {\n",
    "      profile_colors[[unique(dbObj.analyzed$samples$Condition)[i]]] <- c('white', colours[[i]])\n",
    "    }\n",
    "    #profile_colors <- rev(profile_colors)\n",
    "    profiles <- dba.plotProfile(dbObj.analyzed, merge=c(DBA_REPLICATE), normalize=TRUE)\n",
    "    png(paste(output_prefix, 'analyzed_profile.png', sep=''))\n",
    "    #grid.newpage()\n",
    "    g <- dba.plotProfile(profiles, matrices_color=profile_colors, all_color_scales_equal=FALSE, \n",
    "                         decreasing=FALSE, \n",
    "                         #group_anno_color=unlist(unname(conditions_colour_code))\n",
    "    )\n",
    "    #grid.arrange(gTree(children=List(g)), top=textGrob(\"Heatmap-Profiles\", gp=gpar(cex=1.5, fontface='bold')),\n",
    "    #             bottom=textGrob(m, gp=gpar(cex=1.5)))\n",
    "    invisible(capture.output(dev.off()))\n",
    "  }, error=function(e){\n",
    "    message(\"No figure\\n\", e)\n",
    "    invisible(capture.output(dev.off()))\n",
    "    if (file.exists(paste(output_prefix, 'analyzed_profile.png', sep=''))){\n",
    "      invisible(file.remove(paste(output_prefix, 'analyzed_profile.png', sep='')))\n",
    "    }\n",
    "    \n",
    "  }\n",
    ")\n",
    "invisible(capture.output(gc()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b278e69d",
   "metadata": {},
   "source": [
    "### Generate reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a322ee",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Reports for DESeq2 and edgeR, including all peaks\n",
    "reports <- list()\n",
    "reports[[\"DESeq2\"]] <- dba.report(dbObj.analyzed, method=DBA_DESEQ2, contrast=1, th=1)\n",
    "reports[[\"edgeR\"]] <- dba.report(dbObj.analyzed, method=DBA_EDGER, contrast=1, th=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "298f48d0",
   "metadata": {},
   "source": [
    "### Output results, plots, and annotations for both reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5b39d18",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for (report in names(reports)){\n",
    "  \n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  \n",
    "  # Write complete report to file\n",
    "  res <- as.data.frame(reports[[report]])\n",
    "  names(res)[names(res) == 'Fold'] <- 'log2FoldChange'\n",
    "  names(res)[names(res) == 'FDR'] <- 'p.adjust'\n",
    "  res <- res[order(res$log2FoldChange, decreasing=TRUE), ]\n",
    "  res <- as.data.frame(annotatePeak(GRanges(res), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                    level=opt$annotation_level,\n",
    "                                    tssRegion=c(-3000, 3000))@anno)\n",
    "  \n",
    "  write.table(res, file=paste(output_prefix, 'analyzed_report_', report, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  \n",
    "  if (is.null(dba.report(dbObj.analyzed, method=report, contrast=1, th=opt$fdr))){\n",
    "    cat(\"\\nNo DEGs identified by\", report, \"at a significance threshold of\", opt$fdr, \"skipping further analysis...\\n\")\n",
    "    next\n",
    "  }\n",
    "  \n",
    "  # Report columns are seqnames, start, end, width, strand, Conc, Conc_Group1, Conc_Group2, Fold, p.value, FDR\n",
    "  # Create bed files for each keeping only significant peaks (p<0.05)\n",
    "  # Comparing those whereby Fold > 0 vs Fold < 0, indicating enrichment gain vs loss of group 1 over group 2\n",
    "  \n",
    "  out <- as.data.frame(reports[[report]])\n",
    "  #out <- out[as.data.frame(findOverlaps(both, reports[[report]]))[[\"subjectHits\"]], ]\n",
    "  \n",
    "  cat(\"\\n\", report, \"Report:\\n\")\n",
    "  print(head(out))\n",
    "  print(tail(out))\n",
    "  \n",
    "  gained <- out %>% \n",
    "    dplyr::filter(FDR < opt$fdr & Fold > opt$lfc) %>% \n",
    "    dplyr::select(seqnames, start, end)\n",
    "  lost <- out %>% \n",
    "    dplyr::filter(FDR < opt$fdr & Fold < (0 - opt$lfc)) %>% \n",
    "    dplyr::select(seqnames, start, end)\n",
    "  \n",
    "  cat('\\n', dim(out)[[1]], 'peaks in', report, 'report:\\n')\n",
    "  cat('\\n\\t', dim(gained)[[1]] + dim(lost)[[1]], 'statistically significant peaks', '( FDR <', dbObj.analyzed$config$th, ')\\n')\n",
    "  cat('\\n\\t', dim(gained)[[1]], 'DE peaks in', dbObj.contrast$contrasts[[1]]$name1, '(log2Fold-change > ', opt$lfc, '\\n')\n",
    "  cat('\\n\\t', dim(lost)[[1]], 'DE peaks in', dbObj.contrast$contrasts[[1]]$name2, '(log2Fold-change < -', opt$lfc, '\\n')\n",
    "  \n",
    "  \n",
    "  # Write DE result to bed files\n",
    "  if (dim(gained)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name1)\n",
    "    write.table(gained, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name1, '.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=F)\n",
    "  }\n",
    "  if (dim(lost)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name2)\n",
    "    write.table(lost, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name2, '.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=F)\n",
    "  }\n",
    "  \n",
    "  # Plot profile heatmaps for all significant (FDR < opt$fdr) sites for each method\n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  tryCatch(\n",
    "    {\n",
    "      profile_colors <- list()\n",
    "      for (i in 1:length(unique(dbObj.analyzed$samples$Condition))) {\n",
    "        profile_colors[[unique(dbObj.analyzed$samples$Condition)[i]]] <- c('white', colours[[i]])\n",
    "      }\n",
    "      #profile_colors <- rev(profile_colors)\n",
    "      # Plots all significant sites among both conditions (will include signals)\n",
    "      profiles_significant <- dba.plotProfile(dbObj.analyzed, merge=c(DBA_REPLICATE), normalize=TRUE, \n",
    "                                              sites=GRanges(out %>% dplyr::filter(FDR < opt$fdr))\n",
    "      )\n",
    "      png(paste(output_prefix, 'analyzed_profiles_', report,'.png', sep=''))\n",
    "      dba.plotProfile(profiles_significant, matrices_color=profile_colors, all_color_scales_equal=FALSE, \n",
    "                      decreasing=FALSE, \n",
    "                      #group_anno_color=unlist(unname(conditions_colour_code))\n",
    "      )\n",
    "      invisible(capture.output(dev.off()))\n",
    "    }, error=function(e){\n",
    "      message(\"No figure\\n\", e)\n",
    "      invisible(capture.output(dev.off()))\n",
    "      if (file.exists(paste(output_prefix, 'analyzed_profiles_', report,'.png', sep=''))){\n",
    "        invisible(file.remove(paste(output_prefix, 'analyzed_profiles_', report,'.png', sep='')))\n",
    "      }\n",
    "      \n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "  gained <- out %>% \n",
    "    dplyr::filter(FDR < opt$fdr & Fold > opt$lfc)\n",
    "  lost <- out %>% \n",
    "    dplyr::filter(FDR < opt$fdr & Fold < (0 - opt$lfc))\n",
    "  \n",
    "  # Write complete DE result to files\n",
    "  if (dim(gained)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name1)\n",
    "    gained <- as.data.frame(annotatePeak(GRanges(gained), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                         level=opt$annotation_level,\n",
    "                                         tssRegion=c(-3000, 3000))@anno)\n",
    "    gained <- gained[order(gained$Fold, decreasing=TRUE), ]\n",
    "    write.table(gained, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name1, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  }\n",
    "  if (dim(lost)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name2)\n",
    "    lost <- as.data.frame(annotatePeak(GRanges(lost), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                       level=opt$annotation_level,\n",
    "                                       tssRegion=c(-3000, 3000))@anno)\n",
    "    lost <- lost[order(lost$Fold, decreasing=FALSE), ]\n",
    "    write.table(lost, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name2, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  }\n",
    "  \n",
    "  # Use absolute fold-change for plotting magnitude\n",
    "  gained$Fold <- abs(gained$Fold)\n",
    "  lost$Fold <- abs(lost$Fold)\n",
    "  \n",
    "  gpeaks <- GenomicRanges::GRangesList(Gained=gained, Lost=lost)\n",
    "  names(gpeaks) <- c(dbObj.contrast$contrasts[[1]]$name1, dbObj.contrast$contrasts[[1]]$name2)\n",
    "  \n",
    "  # Plot peaks gained/lost over genome between conditions\n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- covplot(gpeaks, title=paste(\"Peaks over Genome\", report, sep=' - '), weightCol='Fold') + \n",
    "        scale_color_manual(values=rev(c(colours[1:length(unique_peaks)]))) + \n",
    "        scale_fill_manual(values=rev(c(colours[1:length(unique_peaks)])))\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_significant_merged_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "      plt <- plt + facet_grid(chr ~ .id)\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_significant_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "    }, error=function(e){\n",
    "      message(\"No figure\\n\", e)\n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "  # ========= Get Annotations =========\n",
    "  peakAnnoList <- list()\n",
    "  for (p in names(gpeaks)){\n",
    "    cat(\"\\nObtaining annotations for \", p, \"\\n\")\n",
    "    output_prefix <- change_dirs(result_dir, report, p)\n",
    "    \n",
    "    if (length(gpeaks[[p]]) == 0){\n",
    "      cat(\"\\nNo\", p, \"peaks to annotate...\\n\")\n",
    "      next\n",
    "    }\n",
    "    \n",
    "    if (p == \"Shared\"){\n",
    "      colour <- \"#56B1F7\"\n",
    "      heat_colour <- \"BuPu\"\n",
    "    }else{\n",
    "      colour <- conditions_colour_code[[p]]\n",
    "    }\n",
    "    \n",
    "    if (colour == \"#00BFC4\"){\n",
    "      heat_colour <- \"Blues\"\n",
    "    }else if (colour == \"#F8766D\"){\n",
    "      heat_colour <- \"Reds\"\n",
    "    }else if (colour == \"#7CAE00\"){\n",
    "      heat_colour <- \"Greens\"\n",
    "    }else if (colour == \"#C77CFF\"){\n",
    "      heat_colour <- \"BuPu\"\n",
    "    }else if (colour == \"#e69e02\"){\n",
    "      heat_colour <- \"YlOrBr\"\n",
    "    }else{\n",
    "      heat_colour <- \"Greys\"\n",
    "    }\n",
    "    \n",
    "    if (length(gpeaks[[p]]) > 0){\n",
    "      cat(\"\\nAnnotating\", p, \"\\n\")\n",
    "      anno <- annotatePeak(gpeaks[[p]], \n",
    "                           TxDb=anno_ref$txdb,\n",
    "                           annoDb=anno_ref$annoDb,\n",
    "                           level=opt$annotation_level,\n",
    "                           tssRegion=c(-3000, 3000))\n",
    "      peakAnnoList[[p]] <- anno\n",
    "      \n",
    "      tryCatch(\n",
    "        {\n",
    "          png(paste(output_prefix,'_', report, 'peaks_annotation_pie', p, '.png', sep=''), width=1680, height=1200)\n",
    "          plt <- plotAnnoPie(anno, main=paste(p, '\\n\\n', length(anno@anno), ' Sites', sep=''), line=-10, cex.main=3.25, cex=3)\n",
    "          invisible(capture.output( dev.off() ))\n",
    "          #invisible(capture.output( ggsave(filename=paste(output_prefix, report, 'peaks_annotation_pie', p, '.png', sep=''), plot=grid.arrange(plt)) ))\n",
    "        },error = function(e)\n",
    "        {\n",
    "          message(e)\n",
    "        }\n",
    "      )\n",
    "      \n",
    "      plt <- upsetplot(anno, vennpie=TRUE) + ggtitle(p)\n",
    "      invisible(capture.output(ggsave(paste(output_prefix, report, '_', p, '_annotated_peaks_upsetplot.png', sep=''), plot=plt, dpi=320, bg='white')))\n",
    "      \n",
    "      # Write annotation to file\n",
    "      write.table(anno, file=paste(output_prefix, 'annotated_', report, '_', p, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "      \n",
    "      tryCatch(\n",
    "        {\n",
    "          genes <- list()\n",
    "          genes[[p]] <- anno@anno$geneId\n",
    "          names(genes) = sub(\"_\", \"\\n\", names(genes))\n",
    "          cat(\"\\nGetting \", report, ' ', p, ' KEGG\\n')\n",
    "          \n",
    "          # fun is \"groupGO\", \"enrichGO\", \"enrichKEGG\", \"enrichDO\" or \"enrichPathway\" \n",
    "          compKEGG <- compareCluster(geneCluster=genes,\n",
    "                                     fun=\"enrichKEGG\",\n",
    "                                     pvalueCutoff=opt$fdr,\n",
    "                                     pAdjustMethod=\"BH\",\n",
    "                                     minGSSize = opt$minGSSize,\n",
    "                                     maxGSSize = opt$maxGSSize,\n",
    "                                     organism=anno_ref$keggOrg\n",
    "          ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "          if (!class(compKEGG) == 'compareClusterResult'){\n",
    "            cat(\"\\nNo results.\\n\")\n",
    "            #next\n",
    "          }else{\n",
    "            cat('\\n', dim(compKEGG@compareClusterResult)[1], 'results\\n')\n",
    "          }\n",
    "          if ((!is.null(compKEGG)) & (dim(compKEGG@compareClusterResult)[1] > 0)){\n",
    "            # Map EntrezIDs to gene SYMBOL\n",
    "            compKEGG@compareClusterResult$SYMBOL <- compKEGG@compareClusterResult$geneID\n",
    "            myEntrez <- lapply(compKEGG@compareClusterResult$geneID, strsplit, '/')\n",
    "            for (i in 1:length(myEntrez)){\n",
    "              compKEGG@compareClusterResult$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "            }\n",
    "            \n",
    "            plt <- make_dotplot(compKEGG@compareClusterResult, title=paste('KEGG - ', p, sep=\"\"), ylabel=\"KEGG Category\", colour=colour, n=15)\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_', p, '_annotated_KEGG.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "            \n",
    "            # Write annotations to csv\n",
    "            write.table(as.data.frame(compKEGG), file=paste(output_prefix, report, '_', p, '_annotated_KEGG.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "            \n",
    "            plt <- make_pheatmapplot(compKEGG@compareClusterResult, res, anno_type=\"KEGG\", assembly=opt$assembly, title=paste('KEGG - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=TRUE, ylabel=\"KEGG Category\")\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_KEGG_annotation_', p, '_pheatmap_bygene.png', sep=''), plot=plt, dpi=320)))\n",
    "            remove(plt)\n",
    "            plt <- make_pheatmapplot(compKEGG@compareClusterResult, res, anno_type=\"KEGG\", assembly=opt$assembly, title=paste('KEGG - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=FALSE, ylabel=\"KEGG Category\")\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_KEGG_annotation_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "            remove(plt)\n",
    "            remove(compKEGG)\n",
    "            gc()\n",
    "          } else{\n",
    "            cat(\"\\nNo annotation results\\n\")\n",
    "            remove(compKEGG)\n",
    "            gc()\n",
    "          }\n",
    "        },error = function(e)\n",
    "        {\n",
    "          message(e)\n",
    "          remove(compKEGG)\n",
    "          gc()\n",
    "        }\n",
    "      )\n",
    "      invisible(capture.output(gc()))\n",
    "      \n",
    "      for (ont in c('ALL', 'CC', 'MF', 'BP')){\n",
    "        cat(\"\\nGetting \", report, ' ',  p, ' GO - ', ont, '\\n')\n",
    "        \n",
    "        tryCatch(\n",
    "          {\n",
    "            compGO <- compareCluster(geneCluster=genes,\n",
    "                                     OrgDb=anno_ref$annoDb,\n",
    "                                     fun=\"enrichGO\",\n",
    "                                     ont=ont,\n",
    "                                     pvalueCutoff=opt$fdr,\n",
    "                                     pAdjustMethod=\"BH\",\n",
    "                                     minGSSize = opt$minGSSize,\n",
    "                                     maxGSSize = opt$maxGSSize,\n",
    "                                     readable=TRUE\n",
    "            ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "            if (!class(compGO) == 'compareClusterResult'){\n",
    "              cat(\"\\nNo results.\\n\")\n",
    "              next\n",
    "            }else{\n",
    "              cat('\\n', dim(compGO@compareClusterResult)[1], 'results\\n')\n",
    "            }\n",
    "            if (!is.null(compGO)){\n",
    "              compGO@compareClusterResult$ONTOLOGY <- go2ont(compGO@compareClusterResult$ID)$Ontology\n",
    "              #plt <- dotplot(compGO, showCategory = 10, title = \"GO Pathway Enrichment Analysis\")\n",
    "              plt <- make_dotplot(compGO@compareClusterResult, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\", colour=colour, n=15)\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_', p, '_annotated_GO-', ont, '.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "              \n",
    "              # Write annotations to csv\n",
    "              write.table(as.data.frame(compGO), file=paste(output_prefix, report, '_', p, '_annotated_GO-', ont, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "              \n",
    "              plt <- make_pheatmapplot(compGO@compareClusterResult, res, heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=TRUE, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\")\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_GO-', ont, '_', p, '_pheatmap_by_gene.png', sep=''), plot=plt, dpi=320)))\n",
    "              remove(plt)\n",
    "              plt <- make_pheatmapplot(compGO@compareClusterResult, res, heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=FALSE, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\")\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_GO-', ont, '_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "              remove(plt)\n",
    "              remove(compGO)\n",
    "              gc()\n",
    "            } else{\n",
    "              cat(\"\\nNo annotation results\\n\")\n",
    "              remove(compGO)\n",
    "              gc()\n",
    "            }\n",
    "          },error = function(e)\n",
    "          {\n",
    "            message(e)\n",
    "            remove(compGO)\n",
    "            gc()\n",
    "          }\n",
    "        )\n",
    "        invisible(capture.output(gc()))\n",
    "        \n",
    "      }\n",
    "      genes_entrez <- list()\n",
    "      genes_entrez[[p]] <- anno@anno$geneId\n",
    "      names(genes_entrez) = sub(\"_\", \"\\n\", names(genes_entrez))\n",
    "      for (annotation_type in c(\"GOTERM_BP_DIRECT\", \"GOTERM_CC_DIRECT\", \"GOTERM_MF_DIRECT\", \"KEGG_PATHWAY\")){\n",
    "        cat(\"\\nGetting \", report, ' ', p, ' DAVID ', annotation_type, ' annotations...\\n')  \n",
    "        # DAVID Annotation\n",
    "        tryCatch(\n",
    "          {\n",
    "            cat(\"\\nDAVID - \", annotation_type, \"\\n\")\n",
    "            entrez <- unname(genes_entrez[[p]][!is.na(unname(genes_entrez[[p]]))])\n",
    "            \n",
    "            py_run_string(\"import pandas as pd\")\n",
    "            py_run_string(\"import sys\")\n",
    "            py_run_string(\"from suds.client import Client\")\n",
    "            \n",
    "            # create a service client using the wsdl.\n",
    "            py_run_string(\"client = Client('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService?wsdl')\")\n",
    "            py_run_string(\"client.wsdl.services[0].setlocation('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService.DAVIDWebServiceHttpSoap11Endpoint/')\")\n",
    "            \n",
    "            #authenticate user email\n",
    "            py_run_string(\"client.service.authenticate(r.opt['david_user'])\")\n",
    "            \n",
    "            # Read input gene list file, convert ids to a comma-delimited string and upload the list to DAVID\n",
    "            py_run_string(\"inputIds = ','.join(r.entrez)\")\n",
    "            py_run_string(\"client.service.addList(inputIds, 'ENTREZ_GENE_ID', r.p, 0)\")\n",
    "            \n",
    "            # setCategories\n",
    "            py_run_string(\"categorySting = str(client.service.setCategories(r.annotation_type))\")\n",
    "            \n",
    "            #getChartReport\n",
    "            py_run_string(\"thd = r.opt['fdr']\")\n",
    "            py_run_string(\"ct = r.opt['minGSSize']\")\n",
    "            py_run_string(\"chartReport = client.service.getChartReport(thd,ct)\")\n",
    "            py_run_string(\"chartRow = len(chartReport)\")\n",
    "            py_run_string(\"print ('Total chart records:',chartRow)\")\n",
    "            \n",
    "            if (py$chartRow > 0){\n",
    "              if (annotation_type == \"KEGG_PATHWAY\"){\n",
    "                splitter <- \":\"\n",
    "              }else{\n",
    "                splitter <- \"~\"\n",
    "              }\n",
    "              # parse chartReport\n",
    "              py_run_string(\"records = pd.DataFrame()\")\n",
    "              py_run_string(\"for simpleChartRecord in chartReport:\n",
    "                              df = pd.DataFrame(index=[records.shape[0]], data={\n",
    "                                  'ID' : simpleChartRecord['termName'].split(r.splitter)[0],\n",
    "                                  'Category' : simpleChartRecord['categoryName'],\n",
    "                                  'Description' : simpleChartRecord['termName'].split(r.splitter)[1],\n",
    "                                  'GeneRatio': str(simpleChartRecord['listHits']) + '/' + str(simpleChartRecord['listTotals']), \n",
    "                                  'BgRatio': str(simpleChartRecord['popHits']) + '/' + str(simpleChartRecord['popTotals']), \n",
    "                                  'pvalue' : simpleChartRecord['ease'],\n",
    "                                  'p.adjust' : simpleChartRecord['benjamini'],\n",
    "                                  'FDR' : simpleChartRecord['afdr'],\n",
    "                                  'geneID' : simpleChartRecord['geneIds'].replace(', ', '/'),\n",
    "                                  'Count' : simpleChartRecord['listHits'],\n",
    "                                  'foldEnrichment' : simpleChartRecord['foldEnrichment'],\n",
    "                                  'id' : simpleChartRecord['id']\n",
    "                                  })\n",
    "                              records = pd.concat([records, df])\")\n",
    "              #py_run_string(\"records.reset_index(drop=True, inplace=True)\")\n",
    "              compD <- py$records\n",
    "              \n",
    "              qobj <- tryCatch(qvalue(p=as.numeric(compD$pvalue), lambda=opt$pvalue, pi0.method=\"bootstrap\"),\n",
    "                               error=function(e) NULL)\n",
    "              if (class(qobj) == \"qvalue\") {\n",
    "                qvalues <- qobj$qvalues\n",
    "              } else {\n",
    "                qvalues <- NA\n",
    "              }\n",
    "              compD$qvalue <- qvalues\n",
    "              \n",
    "              compDAVID <- new(\"enrichResult\",\n",
    "                               result         = compD,\n",
    "                               pvalueCutoff   = 1,\n",
    "                               pAdjustMethod  = \"BH\",\n",
    "                               organism       = opt$assembly,\n",
    "                               ontology       = annotation_type,\n",
    "                               gene           = entrez,\n",
    "                               keytype        = \"ENTREZ_GENE_ID\")\n",
    "              rm(compD)\n",
    "              \n",
    "              if (!class(compDAVID) == 'enrichResult'){\n",
    "                cat(\"\\nNo results.\\n\")\n",
    "                next\n",
    "              }else{\n",
    "                cat('\\n', dim(compDAVID@result)[1], 'results\\n')\n",
    "              }\n",
    "              \n",
    "              if ((!is.null(compDAVID)) & (dim(compDAVID@result)[1] > 0)){\n",
    "                # Map EntrezIDs to gene SYMBOL\n",
    "                compDAVID@result$SYMBOL <- compDAVID@result$geneID\n",
    "                myEntrez <- lapply(compDAVID@result$geneID, strsplit, '/')\n",
    "                for (i in 1:length(myEntrez)){\n",
    "                  compDAVID@result$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "                }\n",
    "                \n",
    "                plt <- make_dotplot(compDAVID@result, title=paste('DAVID - ', p, sep=\"\"), ylabel=paste(annotation_type,\"Category\", sep=' '), colour=colour, n=15)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_dotplot.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                # Write annotations to csv\n",
    "                write.table(as.data.frame(compDAVID), file=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "                \n",
    "                plt <- make_pheatmapplot(compDAVID@result, res, anno_type=\"DAVID\", assembly=opt$assembly, title=paste('DAVID - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=opt$lfc, dendro=TRUE, sort_genes=TRUE)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_pheatmap_bygene.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                plt <- make_pheatmapplot(compDAVID@result, res, anno_type=\"DAVID\", assembly=opt$assembly, title=paste('DAVID - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=opt$lfc, dendro=TRUE, sort_genes=FALSE)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                remove(compDAVID)\n",
    "                gc()\n",
    "              } else{\n",
    "                cat(\"\\nNo annotation results\\n\")\n",
    "                remove(compDAVID)\n",
    "                gc()\n",
    "              }\n",
    "            }\n",
    "          },error = function(e)\n",
    "          {\n",
    "            message(e)\n",
    "            remove(compDAVID)\n",
    "            gc()\n",
    "          }\n",
    "        )\n",
    "      }\n",
    "    }else{\n",
    "        cat(\"\\nNo peaks to annotate for\", p, '\\n')\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    output_prefix <- change_dirs(result_dir, report, '')\n",
    "    tryCatch(\n",
    "      {\n",
    "        plt <- plotAnnoBar(peakAnnoList)\n",
    "        invisible(capture.output(ggsave(filename=paste(output_prefix, 'peaks_annotation_distribution_', report, '.png', sep=''), plot=plt, dpi=320)))\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "      }\n",
    "    )\n",
    "    tryCatch(\n",
    "      {    \n",
    "        plt <- plotDistToTSS(peakAnnoList)\n",
    "        invisible(capture.output(ggsave(filename=paste(output_prefix, 'peaks_annotation_TSS_distribution_', report, '.png', sep=''), plot=plt, dpi=320)))\n",
    "      },error = function(e)\n",
    "      {\n",
    "        message(e)\n",
    "      }\n",
    "    )\n",
    "    invisible(capture.output(gc()))\n",
    "  }\n",
    "  \n",
    "  # Quit if one method produced no results to avoid repeating functions...\n",
    "  if (is.null(dba.report(dbObj.analyzed, method=DBA_DESEQ2, contrast=1, th=opt$fdr)) | is.null(dba.report(dbObj.analyzed, method=DBA_EDGER, contrast=1, th=opt$fdr))){\n",
    "    cat(\"\\nNo DEGs identified by either DESeq2 or edgeR at a significance threshold of\", opt$fdr, \"skipping further analysis...\\n\")\n",
    "    cat(\"\\nFINISHED!\\n\")\n",
    "    q()\n",
    "  }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7507d9a5",
   "metadata": {},
   "source": [
    "### Repeat, isolating for targets identified by both DESeq2 AND edgeR results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d6e624",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81f6f380",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " # For considering sites identified by DESeq2 AND edgeR\n",
    "  #reports[[\"DESeq2_and_edgeR\"]] <- c(reports[[\"DESeq2\"]], reports[[\"edgeR\"]])\n",
    "  #names(reports[[\"DESeq2_and_edgeR\"]]) <- 1:length(names(reports[[\"DESeq2_and_edgeR\"]]))\n",
    "  df_both <- merge(as.data.frame(reports[[\"DESeq2\"]]), as.data.frame(reports[[\"edgeR\"]]), by=c(\"seqnames\", \"start\", \"end\"), suffixes=c(\".DESeq2\", \".edgeR\"))\n",
    "  reports[[\"DESeq2_and_edgeR\"]] <- GRanges(df_both)\n",
    "  report <- \"DESeq2_and_edgeR\"\n",
    "  \n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  \n",
    "  # Write complete report to file\n",
    "  res <- as.data.frame(reports[[report]])\n",
    "  res['log2FoldChange'] <- (res$Fold.DESeq2 + res$Fold.edgeR)/2\n",
    "  res['p.adjust'] <- (res$FDR.DESeq2 + res$FDR.edgeR)/2\n",
    "  res <- as.data.frame(annotatePeak(GRanges(res), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                    level=opt$annotation_level,\n",
    "                                    tssRegion=c(-3000, 3000))@anno)\n",
    "  res <- res[order(res$log2FoldChange, decreasing=TRUE), ]\n",
    "  \n",
    "  write.table(res, file=paste(output_prefix, 'analyzed_report_', report, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  \n",
    "  out <- as.data.frame(reports[[report]])\n",
    "  #out <- out[as.data.frame(findOverlaps(both, reports[[report]]))[[\"subjectHits\"]], ]\n",
    "  \n",
    "  cat(\"\\n\", report, \"Report:\\n\")\n",
    "  print(head(out))\n",
    "  print(tail(out))\n",
    "  \n",
    "  \n",
    "  gained <- out %>% \n",
    "    dplyr::filter(FDR.DESeq2 < opt$fdr & Fold.DESeq2 > opt$lfc & FDR.edgeR < opt$fdr & Fold.edgeR > opt$lfc) %>% \n",
    "    dplyr::select(seqnames, start, end)\n",
    "  lost <- out %>% \n",
    "    dplyr::filter(FDR.DESeq2 < opt$fdr & Fold.DESeq2 < (0 - opt$lfc) & FDR.edgeR < opt$fdr & Fold.edgeR < (0 - opt$lfc)) %>% \n",
    "    dplyr::select(seqnames, start, end)\n",
    "  \n",
    "  cat('\\n', dim(out)[[1]], 'peaks in', report, 'report:\\n')\n",
    "  cat('\\n\\t', dim(gained)[[1]] + dim(lost)[[1]], 'statistically significant peaks', '( FDR <', dbObj.analyzed$config$th, ')\\n')\n",
    "  cat('\\n\\t', dim(gained)[[1]], 'DE peaks log2FoldChange >', opt$lfc, 'in', dbObj.contrast$contrasts[[1]]$name1, '\\n')\n",
    "  cat('\\n\\t', dim(lost)[[1]], 'DE peaks log2FoldChange <', (0 - opt$lfc), 'in', dbObj.contrast$contrasts[[1]]$name2, '\\n')\n",
    "  \n",
    "  \n",
    "  # Write DE result to bed files\n",
    "  if (dim(gained)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name1)\n",
    "    write.table(gained, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name1, '.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=F)\n",
    "  }\n",
    "  if (dim(lost)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name2)\n",
    "    write.table(lost, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name2, '.bed', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=F)\n",
    "  }\n",
    "  \n",
    "  gained <- out %>% \n",
    "    dplyr::filter(FDR.DESeq2 < opt$fdr & Fold.DESeq2 > opt$lfc & FDR.edgeR < opt$fdr & Fold.edgeR > opt$lfc)\n",
    "  lost <- out %>% \n",
    "    dplyr::filter(FDR.DESeq2 < opt$fdr & Fold.DESeq2 < (0 - opt$lfc) & FDR.edgeR < opt$fdr & Fold.edgeR < (0 - opt$lfc))\n",
    "  \n",
    "  \n",
    "  if (dim(gained)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name1)\n",
    "    gained <- as.data.frame(annotatePeak(GRanges(gained), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                         level=opt$annotation_level,\n",
    "                                         tssRegion=c(-3000, 3000))@anno)\n",
    "    gained <- gained[order(gained$Fold.DESeq2, decreasing=TRUE), ]\n",
    "    write.table(gained, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name1, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  }\n",
    "  if (dim(lost)[1] > 0){\n",
    "    output_prefix <- change_dirs(result_dir, report, dbObj.contrast$contrasts[[1]]$name2)\n",
    "    lost <- as.data.frame(annotatePeak(GRanges(lost), TxDb=anno_ref$txdb, annoDb=anno_ref$annoDb,\n",
    "                                       level=opt$annotation_level,\n",
    "                                       tssRegion=c(-3000, 3000))@anno)\n",
    "    lost <- lost[order(lost$Fold.DESeq2, decreasing=FALSE), ]\n",
    "    write.table(lost, file=paste(output_prefix, 'analyzed_report_', report, '_', dbObj.contrast$contrasts[[1]]$name2, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F)\n",
    "  }\n",
    "  \n",
    "  # Use absolute fold-change for plotting magnitude\n",
    "  #gained$Fold <- abs(gained$Fold)\n",
    "  #lost$Fold <- abs(lost$Fold)\n",
    "  \n",
    "  gpeaks <- GenomicRanges::GRangesList(Gained=gained, Lost=lost)\n",
    "  names(gpeaks) <- c(dbObj.contrast$contrasts[[1]]$name1, dbObj.contrast$contrasts[[1]]$name2)\n",
    "  \n",
    "  # Plot peaks gained/lost over genome between conditions\n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- covplot(gpeaks, title=paste(\"Peaks over Genome\", report, sep=' - '), ) + #weightCol='Fold') + \n",
    "        scale_color_manual(values=rev(c(colours[1:length(unique_peaks)]))) + \n",
    "        scale_fill_manual(values=rev(c(colours[1:length(unique_peaks)])))\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_significant_merged_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "      plt <- plt + facet_grid(chr ~ .id)\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_significant_peaks.png', sep=''), plot=plt, dpi=320)))\n",
    "    }, error=function(e){\n",
    "      message(\"No figure\\n\", e)\n",
    "      invisible(capture.output(dev.off()))\n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "  # ========= Get Annotations =========\n",
    "  peakAnnoList <- list()\n",
    "  for (p in names(gpeaks)){\n",
    "    cat(\"\\nObtaining annotations for \", p, \"\\n\")\n",
    "    output_prefix <- change_dirs(result_dir, report, p)\n",
    "    \n",
    "    if (length(gpeaks[[p]]) == 0){\n",
    "      cat(\"\\nNo\", p, \"peaks to annotate...\\n\")\n",
    "      next\n",
    "    }\n",
    "    \n",
    "    if (p == \"Shared\"){\n",
    "      colour <- \"#56B1F7\"\n",
    "      heat_colour <- \"BuPu\"\n",
    "    }else{\n",
    "      colour <- conditions_colour_code[[p]]\n",
    "      heat_colour <- conditions_colour_code[[p]]\n",
    "    }\n",
    "    \n",
    "    if (colour == \"#00BFC4\"){\n",
    "      heat_colour <- \"Blues\"\n",
    "    }else if (colour == \"#F8766D\"){\n",
    "      heat_colour <- \"Reds\"\n",
    "    }else if (colour == \"#7CAE00\"){\n",
    "      heat_colour <- \"Greens\"\n",
    "    }else if (colour == \"#C77CFF\"){\n",
    "      heat_colour <- \"BuPu\"\n",
    "    }else if (colour == \"#e69e02\"){\n",
    "      heat_colour <- \"YlOrBr\"\n",
    "    }else{\n",
    "      heat_colour <- \"Greys\"\n",
    "    }\n",
    "    \n",
    "    if (length(gpeaks[[p]]) > 0){\n",
    "      cat(\"\\nAnnotating\", p, \"\\n\")\n",
    "      anno <- annotatePeak(gpeaks[[p]], \n",
    "                           TxDb=anno_ref$txdb,\n",
    "                           annoDb=anno_ref$annoDb,\n",
    "                           level=opt$annotation_level,\n",
    "                           tssRegion=c(-3000, 3000))\n",
    "      peakAnnoList[[p]] <- anno\n",
    "      \n",
    "      tryCatch(\n",
    "        {\n",
    "          png(paste(output_prefix,'_', report, 'peaks_annotation_pie', p, '.png', sep=''), width=1680, height=1200)\n",
    "          plt <- plotAnnoPie(anno, main=paste(p, '\\n\\n', length(anno@anno), ' Sites', sep=''), line=-10, cex.main=3.25, cex=3)\n",
    "          invisible(capture.output( dev.off() ))\n",
    "          #invisible(capture.output( ggsave(filename=paste(output_prefix, report, 'peaks_annotation_pie', p, '.png', sep=''), plot=grid.arrange(plt)) ))\n",
    "        },error = function(e)\n",
    "        {\n",
    "          message(e)\n",
    "        }\n",
    "      )\n",
    "      \n",
    "      plt <- upsetplot(anno, vennpie=TRUE) + ggtitle(p)\n",
    "      invisible(capture.output(ggsave(paste(output_prefix, report, '_', p, '_annotated_peaks_upsetplot.png', sep=''), plot=plt, dpi=320, bg='white')))\n",
    "      \n",
    "      # Write annotation to file\n",
    "      write.table(anno, file=paste(output_prefix, 'annotated_', report, '_', p, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "      \n",
    "      tryCatch(\n",
    "        {\n",
    "          genes <- list()\n",
    "          genes[[p]] <- anno@anno$geneId\n",
    "          names(genes) = sub(\"_\", \"\\n\", names(genes))\n",
    "          cat(\"\\nGetting\", report, p, 'KEGG\\n')\n",
    "          \n",
    "          # fun is \"groupGO\", \"enrichGO\", \"enrichKEGG\", \"enrichDO\" or \"enrichPathway\" \n",
    "          compKEGG <- compareCluster(geneCluster=genes,\n",
    "                                     fun=\"enrichKEGG\",\n",
    "                                     pvalueCutoff=opt$fdr,\n",
    "                                     pAdjustMethod=\"BH\",\n",
    "                                     minGSSize = opt$minGSSize,\n",
    "                                     maxGSSize = opt$maxGSSize,\n",
    "                                     organism=anno_ref$keggOrg\n",
    "          ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "          if (!class(compKEGG) == 'compareClusterResult'){\n",
    "            cat(\"\\nNo results.\\n\")\n",
    "            #next\n",
    "          }else{\n",
    "            cat('\\n', dim(compKEGG@compareClusterResult)[1], 'results\\n')\n",
    "          }\n",
    "          if ((!is.null(compKEGG)) & (dim(compKEGG@compareClusterResult)[1] > 0)){\n",
    "            # Map EntrezIDs to gene SYMBOL\n",
    "            compKEGG@compareClusterResult$SYMBOL <- compKEGG@compareClusterResult$geneID\n",
    "            myEntrez <- lapply(compKEGG@compareClusterResult$geneID, strsplit, '/')\n",
    "            for (i in 1:length(myEntrez)){\n",
    "              compKEGG@compareClusterResult$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "            }\n",
    "            \n",
    "            plt <- make_dotplot(compKEGG@compareClusterResult, title=paste('KEGG - ', p, sep=\"\"), ylabel=\"KEGG Category\", colour=colour, n=15)\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_', p, '_annotated_KEGG.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "            \n",
    "            # Write annotations to csv\n",
    "            write.table(as.data.frame(compKEGG), file=paste(output_prefix, report, '_', p, '_annotated_KEGG.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "            \n",
    "            plt <- make_pheatmapplot(compKEGG@compareClusterResult, res, anno_type=\"KEGG\", assembly=opt$assembly, title=paste('KEGG - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=TRUE, ylabel=\"KEGG Category\")\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_KEGG_annotation_', p, '_pheatmap_bygene.png', sep=''), plot=plt, dpi=320)))\n",
    "            remove(plt)\n",
    "            plt <- make_pheatmapplot(compKEGG@compareClusterResult, res, anno_type=\"KEGG\", assembly=opt$assembly, title=paste('KEGG - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=FALSE, ylabel=\"KEGG Category\")\n",
    "            invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_KEGG_annotation_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "            remove(plt)\n",
    "            remove(compKEGG)\n",
    "            gc()\n",
    "          } else{\n",
    "            cat(\"\\nNo annotation results\\n\")\n",
    "            remove(compKEGG)\n",
    "            gc()\n",
    "          }\n",
    "        },error = function(e)\n",
    "        {\n",
    "          message(e)\n",
    "          remove(compKEGG)\n",
    "          gc()\n",
    "        }\n",
    "      )\n",
    "      invisible(capture.output(gc()))\n",
    "      \n",
    "      for (ont in c('ALL', 'CC', 'MF', 'BP')){\n",
    "        cat(\"\\nGetting \", report, ' ',  p, ' GO - ', ont, '\\n')\n",
    "        tryCatch(\n",
    "          {\n",
    "            compGO <- compareCluster(geneCluster=genes,\n",
    "                                     OrgDb=anno_ref$annoDb,\n",
    "                                     fun=\"enrichGO\",\n",
    "                                     ont=ont,\n",
    "                                     pvalueCutoff=opt$fdr,\n",
    "                                     pAdjustMethod=\"BH\",\n",
    "                                     minGSSize = opt$minGSSize,\n",
    "                                     maxGSSize = opt$maxGSSize,\n",
    "                                     readable=TRUE\n",
    "            ) # Check https://www.genome.jp/kegg/catalog/org_list.html for organism hsa=human mmu=mouse\n",
    "            if (!class(compGO) == 'compareClusterResult'){\n",
    "              cat(\"\\nNo results.\\n\")\n",
    "              next\n",
    "            }else{\n",
    "              cat('\\n', dim(compGO@compareClusterResult)[1], 'results\\n')\n",
    "            }\n",
    "            if ((!is.null(compGO)) & (dim(compGO@compareClusterResult)[1] > 0)){\n",
    "              compGO@compareClusterResult$ONTOLOGY <- go2ont(compGO@compareClusterResult$ID)$Ontology\n",
    "              #plt <- dotplot(compGO, showCategory = 10, title = \"GO Pathway Enrichment Analysis\")\n",
    "              plt <- make_dotplot(compGO@compareClusterResult, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\", colour=colour, n=15)\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_', p, '_annotated_GO-', ont, '.png', sep=''), plot=plt, dpi=320, width=10, units='in')))\n",
    "              \n",
    "              # Write annotations to csv\n",
    "              write.table(as.data.frame(compGO), file=paste(output_prefix, report, '_', p, '_annotated_GO-', ont, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "              \n",
    "              plt <- make_pheatmapplot(compGO@compareClusterResult, res, assembly=opt$assembly, heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=TRUE, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\")\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_GO-', ont, '_', p, '_pheatmap_by_gene.png', sep=''), plot=plt, dpi=320)))\n",
    "              remove(plt)\n",
    "              plt <- make_pheatmapplot(compGO@compareClusterResult, res, assembly=opt$assembly, heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=round(max(res$log2FoldChange)), dendro=TRUE, sort_genes=FALSE, title=paste(\"GO (\", ont, \") - \", p, sep=\"\"), ylabel=\"GO Term\")\n",
    "              invisible(capture.output(ggsave(filename=paste(output_prefix, report, '_GO-', ont, '_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "              remove(plt)\n",
    "              remove(compGO)\n",
    "              gc()\n",
    "            } else{\n",
    "              cat(\"\\nNo annotation results\\n\")\n",
    "              remove(compGO)\n",
    "              gc()\n",
    "            }\n",
    "          },error = function(e)\n",
    "          {\n",
    "            message(e)\n",
    "            remove(compGO)\n",
    "            gc()\n",
    "          }\n",
    "        )\n",
    "        invisible(capture.output(gc()))\n",
    "        \n",
    "      }\n",
    "      \n",
    "      genes_entrez <- list()\n",
    "      genes_entrez[[p]] <- anno@anno$geneId\n",
    "      names(genes_entrez) = sub(\"_\", \"\\n\", names(genes_entrez))\n",
    "      for (annotation_type in c(\"GOTERM_BP_DIRECT\", \"GOTERM_CC_DIRECT\", \"GOTERM_MF_DIRECT\", \"KEGG_PATHWAY\")){\n",
    "        cat(\"\\nGetting \", report, ' ', p, ' DAVID ', annotation_type, ' annotations...\\n')  \n",
    "        # DAVID Annotation\n",
    "        tryCatch(\n",
    "          {\n",
    "            cat(\"\\nDAVID - \", annotation_type, \"\\n\")\n",
    "            entrez <- unname(genes_entrez[[p]][!is.na(unname(genes_entrez[[p]]))])\n",
    "            \n",
    "            py_run_string(\"import pandas as pd\")\n",
    "            py_run_string(\"import sys\")\n",
    "            py_run_string(\"from suds.client import Client\")\n",
    "            \n",
    "            # create a service client using the wsdl.\n",
    "            py_run_string(\"client = Client('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService?wsdl')\")\n",
    "            py_run_string(\"client.wsdl.services[0].setlocation('https://davidbioinformatics.nih.gov/webservice/services/DAVIDWebService.DAVIDWebServiceHttpSoap11Endpoint/')\")\n",
    "            \n",
    "            #authenticate user email\n",
    "            py_run_string(\"client.service.authenticate(r.opt['david_user'])\")\n",
    "            \n",
    "            # Read input gene list file, convert ids to a comma-delimited string and upload the list to DAVID\n",
    "            py_run_string(\"inputIds = ','.join(r.entrez)\")\n",
    "            py_run_string(\"client.service.addList(inputIds, 'ENTREZ_GENE_ID', r.p, 0)\")\n",
    "            \n",
    "            # setCategories\n",
    "            py_run_string(\"categorySting = str(client.service.setCategories(r.annotation_type))\")\n",
    "            \n",
    "            #getChartReport\n",
    "            py_run_string(\"thd = r.opt['fdr']\")\n",
    "            py_run_string(\"ct = r.opt['minGSSize']\")\n",
    "            py_run_string(\"chartReport = client.service.getChartReport(thd,ct)\")\n",
    "            py_run_string(\"chartRow = len(chartReport)\")\n",
    "            py_run_string(\"print ('Total chart records:',chartRow)\")\n",
    "            \n",
    "            if (py$chartRow > 0){\n",
    "              if (annotation_type == \"KEGG_PATHWAY\"){\n",
    "                splitter <- \":\"\n",
    "              }else{\n",
    "                splitter <- \"~\"\n",
    "              }\n",
    "              # parse chartReport\n",
    "              py_run_string(\"records = pd.DataFrame()\")\n",
    "              py_run_string(\"for simpleChartRecord in chartReport:\n",
    "                          df = pd.DataFrame(index=[records.shape[0]], data={\n",
    "                              'ID' : simpleChartRecord['termName'].split(r.splitter)[0],\n",
    "                              'Category' : simpleChartRecord['categoryName'],\n",
    "                              'Description' : simpleChartRecord['termName'].split(r.splitter)[1],\n",
    "                              'GeneRatio': str(simpleChartRecord['listHits']) + '/' + str(simpleChartRecord['listTotals']), \n",
    "                              'BgRatio': str(simpleChartRecord['popHits']) + '/' + str(simpleChartRecord['popTotals']), \n",
    "                              'pvalue' : simpleChartRecord['ease'],\n",
    "                              'p.adjust' : simpleChartRecord['benjamini'],\n",
    "                              'FDR' : simpleChartRecord['afdr'],\n",
    "                              'geneID' : simpleChartRecord['geneIds'].replace(', ', '/'),\n",
    "                              'Count' : simpleChartRecord['listHits'],\n",
    "                              'foldEnrichment' : simpleChartRecord['foldEnrichment'],\n",
    "                              'id' : simpleChartRecord['id']\n",
    "                              })\n",
    "                          records = pd.concat([records, df])\")\n",
    "              #py_run_string(\"records.reset_index(drop=True, inplace=True)\")\n",
    "              compD <- py$records\n",
    "              \n",
    "              qobj <- tryCatch(qvalue(p=as.numeric(compD$pvalue), lambda=opt$pvalue, pi0.method=\"bootstrap\"),\n",
    "                               error=function(e) NULL)\n",
    "              if (class(qobj) == \"qvalue\") {\n",
    "                qvalues <- qobj$qvalues\n",
    "              } else {\n",
    "                qvalues <- NA\n",
    "              }\n",
    "              compD$qvalue <- qvalues\n",
    "              \n",
    "              compDAVID <- new(\"enrichResult\",\n",
    "                               result         = compD,\n",
    "                               pvalueCutoff   = 1,\n",
    "                               pAdjustMethod  = \"BH\",\n",
    "                               organism       = opt$assembly,\n",
    "                               ontology       = annotation_type,\n",
    "                               gene           = entrez,\n",
    "                               keytype        = \"ENTREZ_GENE_ID\")\n",
    "              rm(compD)\n",
    "              \n",
    "              if (!class(compDAVID) == 'enrichResult'){\n",
    "                cat(\"\\nNo results.\\n\")\n",
    "                next\n",
    "              }else{\n",
    "                cat('\\n', dim(compDAVID@result)[1], 'results\\n')\n",
    "              }\n",
    "              \n",
    "              if ((!is.null(compDAVID)) & (dim(compDAVID@result)[1] > 0)){\n",
    "                # Map EntrezIDs to gene SYMBOL\n",
    "                compDAVID@result$SYMBOL <- compDAVID@result$geneID\n",
    "                myEntrez <- lapply(compDAVID@result$geneID, strsplit, '/')\n",
    "                for (i in 1:length(myEntrez)){\n",
    "                  compDAVID@result$SYMBOL[i] <- paste(plyr::mapvalues(myEntrez[[i]][[1]], mapper$geneId, mapper$SYMBOL, warn_missing = FALSE), collapse='/')\n",
    "                }\n",
    "                \n",
    "                plt <- make_dotplot(compDAVID@result, title=paste('DAVID - ', p, sep=\"\"), ylabel=paste(annotation_type,\"Category\", sep=' '), colour=colour, n=15)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_dotplot.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                # Write annotations to csv\n",
    "                write.table(as.data.frame(compDAVID), file=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '.tsv', sep=''), sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "                \n",
    "                plt <- make_pheatmapplot(compDAVID@result, res, anno_type=\"DAVID\", assembly=opt$assembly, title=paste('DAVID - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=opt$lfc, dendro=TRUE, sort_genes=TRUE)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_pheatmap_bygene.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                plt <- make_pheatmapplot(compDAVID@result, res, anno_type=\"DAVID\", assembly=opt$assembly, title=paste('DAVID - ', p, sep=\"\"), heat_colour = heat_colour, num_terms=25, num_genes=50, lfc=opt$lfc, dendro=TRUE, sort_genes=FALSE)\n",
    "                invisible(capture.output(ggsave(filename=paste(output_prefix, report, 'DAVID_annotation_', annotation_type, '_', p, '_pheatmap.png', sep=''), plot=plt, dpi=320)))\n",
    "                remove(plt)\n",
    "                remove(compDAVID)\n",
    "                gc()\n",
    "              } else{\n",
    "                cat(\"\\nNo annotation results\\n\")\n",
    "                remove(compDAVID)\n",
    "                gc()\n",
    "              }\n",
    "            }\n",
    "            \n",
    "          },error = function(e)\n",
    "          {\n",
    "            message(e)\n",
    "            remove(compDAVID)\n",
    "            gc()\n",
    "          }\n",
    "        )\n",
    "      }\n",
    "        \n",
    "    }else{\n",
    "        cat(\"No peaks to annotate for\", p, '\\n')\n",
    "      }\n",
    "    }\n",
    "    \n",
    "  output_prefix <- change_dirs(result_dir, report, '')\n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- plotAnnoBar(peakAnnoList)\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, 'peaks_annotation_distribution_', report, '.png', sep=''), plot=plt, dpi=320)))\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "    }\n",
    "  )\n",
    "  tryCatch(\n",
    "    {\n",
    "      plt <- plotDistToTSS(peakAnnoList)\n",
    "      invisible(capture.output(ggsave(filename=paste(output_prefix, 'peaks_annotation_TSS_distribution_', report, '.png', sep=''), plot=plt, dpi=320)))\n",
    "    },error = function(e)\n",
    "    {\n",
    "      message(e)\n",
    "    }\n",
    "  )\n",
    "  invisible(capture.output(gc()))\n",
    "  \n",
    "  cat('\\nFINISHED!\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d367e0",
   "metadata": {},
   "source": [
    "## <center> =========== END OF Affinity Analysis ============ </center>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
